

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Testing Evolutionary Hypothesis &mdash; ETE Toolkit - analysis and visualization of trees</title>
  

  
  
    <link rel="shortcut icon" href="../_static/favicon.ico"/>
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  

  

  
        <link rel="author" title="About these documents"
              href="../about.html"/>
    <link rel="top" title="ETE Toolkit - analysis and visualization of trees" href="../index.html"/>
        <link rel="up" title="The ETE tutorial" href="index.html"/>
        <link rel="next" title="Dealing with the NCBI Taxonomy database" href="tutorial_ncbitaxonomy.html"/>
        <link rel="prev" title="Interactive web tree visualization" href="tutorial_webplugin.html"/>
<script>
    (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
           (i[r].q=i[r].q||[]).push(arguments)
    },i[r].l=1*new Date();a=s.createElement(o),
    m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
    })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

 ga('create', 'UA-783841-4', 'etetoolkit.org');
 ga('send', 'pageview');
 </script>

<!-- 
 <script type="text/javascript">
 var _gaq = _gaq || [];
 _gaq.push(['_setAccount', 'UA-783841-12']);
 _gaq.push(['_setCookiePath', '/ete2/']);
 _gaq.push(['_trackPageview']);

 (function() {
 var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
 ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
 var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
 })();
 </script>
 -->
<!-- <link rel="canonical" href="http://etetoolkit.org/docs/2.3/CANONICALPATH" />
<meta http-equiv="refresh" content="0; url=http:/etetoolkit.org/CANONICALPATH" />
<script type="text/javascript">
window.location.replace('http://etetoolkit.org/docs/2.3/' + window.location.pathname.replace('/ete2/',''));
</script> -->
 
 


  
  <script src="../_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="../index.html" class="icon icon-home"> ETE Toolkit
          

          
          </a>

          
            
            
              <div class="version">
                3.0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
                <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../changelog/index.html">Changelog history</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">The ETE tutorial</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="tutorial_trees.html">Working With Tree Data Structures</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial_drawing.html">The Programmable Tree Drawing Engine</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial_phylogeny.html">Phylogenetic Trees</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial_clustering.html">Clustering Trees</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial_xml.html">Phylogenetic XML standards</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial_webplugin.html">Interactive web tree visualization</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="">Testing Evolutionary Hypothesis</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#extension-requirements">Extension Requirements:</a></li>
<li class="toctree-l3"><a class="reference internal" href="#overview">Overview</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#the-working-directory">The working directory</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#descriptive-analysis">Descriptive analysis</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#branch-model">Branch model</a></li>
<li class="toctree-l4"><a class="reference internal" href="#site-model">Site model</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#hypothesis-testing">Hypothesis Testing</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#test-on-sites">Test on sites</a></li>
<li class="toctree-l4"><a class="reference internal" href="#test-on-branches">Test on branches</a></li>
<li class="toctree-l4"><a class="reference internal" href="#test-on-branch-sites">Test on branch-sites</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#utilities">Utilities</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#load-precomputed-evolutionary-model">Load precomputed evolutionary model</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#references">References</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="tutorial_ncbitaxonomy.html">Dealing with the NCBI Taxonomy database</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial_etree2orthoxml.html">SCRIPTS: orthoXML</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../reference/index.html">ETE&#8217;s Reference Guide</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="../index.html">ETE Toolkit</a>
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          





<div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="../index.html">Docs</a> &raquo;</li>
      
          <li><a href="index.html">The ETE tutorial</a> &raquo;</li>
      
    <li>Testing Evolutionary Hypothesis</li>
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="../_sources/tutorial/tutorial_adaptation.txt" rel="nofollow"> View page source</a>
          
        
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="testing-evolutionary-hypothesis">
<h1><a class="toc-backref" href="#id4">Testing Evolutionary Hypothesis</a><a class="headerlink" href="#testing-evolutionary-hypothesis" title="Permalink to this headline">¶</a></h1>
<div class="contents topic" id="contents">
<p class="topic-title first">Contents</p>
<ul class="simple">
<li><a class="reference internal" href="#testing-evolutionary-hypothesis" id="id4">Testing Evolutionary Hypothesis</a><ul>
<li><a class="reference internal" href="#extension-requirements" id="id5">Extension Requirements:</a></li>
<li><a class="reference internal" href="#overview" id="id6">Overview</a><ul>
<li><a class="reference internal" href="#the-working-directory" id="id7">The working directory</a></li>
</ul>
</li>
<li><a class="reference internal" href="#descriptive-analysis" id="id8">Descriptive analysis</a><ul>
<li><a class="reference internal" href="#branch-model" id="id9">Branch model</a></li>
<li><a class="reference internal" href="#site-model" id="id10">Site model</a></li>
</ul>
</li>
<li><a class="reference internal" href="#hypothesis-testing" id="id11">Hypothesis Testing</a><ul>
<li><a class="reference internal" href="#test-on-sites" id="id12">Test on sites</a></li>
<li><a class="reference internal" href="#test-on-branches" id="id13">Test on branches</a></li>
<li><a class="reference internal" href="#test-on-branch-sites" id="id14">Test on branch-sites</a></li>
</ul>
</li>
<li><a class="reference internal" href="#utilities" id="id15">Utilities</a><ul>
<li><a class="reference internal" href="#load-precomputed-evolutionary-model" id="id16">Load precomputed evolutionary model</a></li>
</ul>
</li>
<li><a class="reference internal" href="#references" id="id17">References</a></li>
</ul>
</li>
</ul>
</div>
<div class="section" id="extension-requirements">
<h2><a class="toc-backref" href="#id5">Extension Requirements:</a><a class="headerlink" href="#extension-requirements" title="Permalink to this headline">¶</a></h2>
<p>You should have codeml and slr in your path:</p>
<blockquote>
<div><ul class="simple">
<li>CodeML, you can download it from
<a class="reference external" href="http://abacus.gene.ucl.ac.uk/software/paml.html">http://abacus.gene.ucl.ac.uk/software/paml.html</a></li>
<li>SLR, from here: <a class="reference external" href="http://www.ebi.ac.uk/goldman-srv/SLR/">http://www.ebi.ac.uk/goldman-srv/SLR/</a></li>
</ul>
</div></blockquote>
<p>Download, compile and install both of the programs, in order to be able to run the examples.</p>
<p>This ETE extension is written on top of these two external programs, so it is strongly recommended to read their corresponding documentation.</p>
</div>
<div class="section" id="overview">
<h2><a class="toc-backref" href="#id6">Overview</a><a class="headerlink" href="#overview" title="Permalink to this headline">¶</a></h2>
<p>An other aspect in the study of evolutionary history, is the analysis of selective pressures accounting for the conservation or degeneration of <strong>protein coding genes</strong>.</p>
<p>The <a class="reference internal" href="../reference/reference_evoltree.html#ete3.EvolTree" title="ete3.EvolTree"><code class="xref py py-class docutils literal"><span class="pre">EvolTree</span></code></a> class is an extension of the class <a class="reference internal" href="../reference/reference_phylo.html#ete3.PhyloTree" title="ete3.PhyloTree"><code class="xref py py-class docutils literal"><span class="pre">PhyloTree</span></code></a> that implements mainly bindings to the PAML package <a class="reference internal" href="#yang2007" id="id1">[yang2007]</a> but also to the SLR program <a class="reference internal" href="#massingham2005" id="id2">[massingham2005]</a>.</p>
<p>Evolutionary variables that are used to summary selective pressures are, of course the branch-length (<em>bL</em>) already available in <a class="reference internal" href="../reference/reference_phylo.html#ete3.PhyloTree" title="ete3.PhyloTree"><code class="xref py py-class docutils literal"><span class="pre">PhyloTree</span></code></a>, but also the rate of non-synonymous mutations (<em>dN</em>), the rate of synonymous mutations (<em>dS</em>) and finally the <span class="math">\(\omega\)</span> ratio:</p>
<div class="math">
\[\begin{eqnarray}
  \omega = \frac{dN}{dS}
\end{eqnarray}\]</div><div class="section" id="the-working-directory">
<h3><a class="toc-backref" href="#id7">The working directory</a><a class="headerlink" href="#the-working-directory" title="Permalink to this headline">¶</a></h3>
<p>EvolTree works mainly as PhyloTree, thus it needs a tree and an alignment. However as you are going to run external programs over it, <strong>a working directory needs to be defined</strong>. By default tree.workdiris &#8220;/tmp/ete3-codeml/&#8221;, but it is recommended to change it to a more useful path.</p>
<p>Jobs will be stored in the workdir, and you will be able to load precomputed evolutionary models from there.</p>
</div>
</div>
<div class="section" id="descriptive-analysis">
<h2><a class="toc-backref" href="#id8">Descriptive analysis</a><a class="headerlink" href="#descriptive-analysis" title="Permalink to this headline">¶</a></h2>
<p>In order to identify the evolutionary trends in a phylogenetic tree, one can either:</p>
<blockquote>
<div><ul class="simple">
<li>conduct an analysis over branches and compute the value of
<span class="math">\(\omega\)</span> in each of them.</li>
<li>look at the selective pressures along the alignment.</li>
</ul>
</div></blockquote>
<div class="section" id="branch-model">
<h3><a class="toc-backref" href="#id9">Branch model</a><a class="headerlink" href="#branch-model" title="Permalink to this headline">¶</a></h3>
<p>As for <a class="reference internal" href="../reference/reference_phylo.html#ete3.PhyloTree" title="ete3.PhyloTree"><code class="xref py py-class docutils literal"><span class="pre">PhyloTree</span></code></a>, we first load the tree and alignment (and you working directory, if you want to save a copy of your jobs):</p>
<div class="highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">ete3</span> <span class="kn">import</span> <span class="n">EvolTree</span>

<span class="n">tree</span> <span class="o">=</span> <span class="n">EvolTree</span><span class="p">(</span><span class="s">&quot;((Hylobates_lar,(Gorilla_gorilla,Pan_troglodytes)),Papio_cynocephalus);&quot;</span><span class="p">)</span>

<span class="n">tree</span><span class="o">.</span><span class="n">link_to_alignment</span> <span class="p">(</span><span class="s">&#39;&#39;&#39;&gt;Hylobates_lar</span>
<span class="s">ATGGCCAGGTACAGATGCTGCCGCAGCCAGAGCCGGAGCAGATGTTACCGCCAGAGCCGGAGCAGATGTTACCGCCAGAGGCAAAGCCAGAGTCGGAGCAGATGTTACCGCCAGAGCCAGAGCCGGAGCAGATGTTACCGCCAGAGACAAAGAAGTCGGAGACGAAGGAGGCGGAGCTGCCAGACACGGAGGAGAGCCATGAGGTGT---CGCCGCAGGTACAGGCTGAGACGTAGAAGCTGTTACCACATTGTATCT</span>
<span class="s">&gt;Papio_cynocephalus</span>
<span class="s">ATGGCCAGGTACAGATGCTGCCGCAGCCAGAGCCGAAGCAGATGCTATCGCCAGAGCCGGAGCAGATGTAACCGCCAGAGACAGAGCCAAAGCCGGAGAAGCTGCTATCGCCAGAGCCAAAGCCGGAGCAGATGTTACCGCCAGAGACAGAGAAGTCGTAGACGAAGGAGGCGACGCTGCCAGACACGGAGGAGAGCCATGAGGTGCTTCCGCCGCAGGTACAGGCTGAGGCGTAGGAGGCCCTATCACATCGTGTCT</span>
<span class="s">&gt;Gorilla_gorilla</span>
<span class="s">ATGGCCAGGTACAGATGCTGTCGCAGCCAGAGCCGCAGCAGATGTTACCGGCAGAGCCGGAGCAGGTGTTACCGGCAGAGACAAAGCCAGAGCCGGAGCAGATGCTACCGGCAGAGCCAAAGCCGGAGCAGGTGTTACCGGCAGAGACAAAGAAGTCGCAGACGTAGGCGGAGGAGCTGCCAGACACGGAGGAGAGCCATGAGGTGCTGCCGCCGCAGGTACAGACTGAGACGTAGAAGACCCTATCATATTGTATCT</span>
<span class="s">&gt;Pan_troglodytes</span>
<span class="s">ATGGCCAGGTACAGATGCTGTCGCAGCCAGAGCCGGAGCAGATGTTACCGGCAGAGACGGAGCAGGTGTTACCGGCAAAGGCAAAGCCAAAGTCGGAGCAGATGTTACCGGCAGAGCCAGAGACGGAGCAGGTGTTACCGGCAAAGACAAAGAAGTCGCAGACGAAGGCGACGGAGCTGCCAGACACGGAGGAGAGCCATGAGGTGCTGCCGCCGCAGGTACAGACTGAGACGTAAAAGATGTTACCATATTGTATCT&#39;&#39;&#39;</span><span class="p">)</span>

<span class="n">tree</span><span class="o">.</span><span class="n">workdir</span> <span class="o">=</span> <span class="s">&#39;/path_to/my_working_directory/&#39;</span>
</pre></div>
</div>
<p>Once loaded we are able to compute selective pressure among the tree according to an evolutionary model. In this case, we will use free-ratio model:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">tree</span><span class="o">.</span><span class="n">run_model</span> <span class="p">(</span><span class="s">&#39;fb.example&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p><a class="reference internal" href="../reference/reference_evoltree.html#ete3.EvolNode.run_model" title="ete3.EvolNode.run_model"><code class="xref py py-func docutils literal"><span class="pre">EvolNode.run_model()</span></code></a> allows to run different evolutionary models (follow this link <a class="reference internal" href="../reference/reference_evoltree.html#ete3.EvolNode.run_model" title="ete3.EvolNode.run_model"><code class="xref py py-func docutils literal"><span class="pre">EvolNode.run_model()</span></code></a> to see the full list of models available). By convention, the name of the model called is the first word, the rest of the string, after the dot, corresponds to its identifier in order to differentiate different runs of one model. We can thus either run &#8220;fb&#8221; model and &#8220;fb.something&#8221; with the same result, but they will be stored under different names instead of being overwritten.</p>
<p>Once computation is done, we can retrieve the <code class="xref py py-class docutils literal"><span class="pre">evol.Model</span></code> object from the tree, with the <a class="reference internal" href="../reference/reference_evoltree.html#ete3.EvolNode.get_evol_model" title="ete3.EvolNode.get_evol_model"><code class="xref py py-func docutils literal"><span class="pre">EvolNode.get_evol_model()</span></code></a> function.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">my_first_fb_model</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">get_evol_model</span><span class="p">(</span><span class="s">&#39;fb.example&#39;</span><span class="p">)</span>

<span class="k">print</span> <span class="n">my_first_fb_model</span>
</pre></div>
</div>
<p>This last print statement would ouptut:</p>
<div class="highlight-python"><div class="highlight"><pre>Evolutionary Model fb.example:
       log likelihood       : -521.421323
       number of parameters : 13
       sites inference      : None
       sites classes        : None
       branches             :
       mark: #0  , omega: None      , node_ids: 5   , name: NoName
       mark: #1  , omega: 0.0001    , node_ids: 6   , name: NoName
       mark: #2  , omega: 999.0     , node_ids: 4   , name: Papio_cynocephalus
       mark: #3  , omega: 999.0     , node_ids: 2   , name: Hylobates_lar
       mark: #4  , omega: 0.0001    , node_ids: 7   , name: NoName
       mark: #5  , omega: 0.1049    , node_ids: 1   , name: Gorilla_gorilla
       mark: #6  , omega: 0.5334    , node_ids: 3   , name: Pan_troglodytes
</pre></div>
</div>
<p>By default, the free-branch model is run with this parameters:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="k">print</span> <span class="n">my_first_fb_model</span><span class="o">.</span><span class="n">get_ctrl_string</span><span class="p">()</span>

<span class="c">#      seqfile = algn</span>
<span class="c">#     treefile = tree</span>
<span class="c">#      outfile = out</span>
<span class="c">#</span>
<span class="c">#       aaDist = 0</span>
<span class="c">#    fix_alpha = 1</span>
<span class="c">#        alpha = 0.0</span>
<span class="c">#  fix_blength = 0</span>
<span class="c">#    cleandata = 0</span>
<span class="c">#        clock = 0</span>
<span class="c">#    CodonFreq = 2</span>
<span class="c">#        getSE = 0</span>
<span class="c">#        icode = 0</span>
<span class="c">#    fix_kappa = 0</span>
<span class="c">#        kappa = 2</span>
<span class="c">#       Malpha = 0</span>
<span class="c">#       method = 0</span>
<span class="c">#        Mgene = 0</span>
<span class="c">#        model = 1</span>
<span class="c">#        ncatG = 8</span>
<span class="c">#        noisy = 0</span>
<span class="c">#      NSsites = 0</span>
<span class="c">#    fix_omega = 0</span>
<span class="c">#        omega = 0.7</span>
<span class="c"># RateAncestor = 0</span>
<span class="c">#      runmode = 0</span>
<span class="c">#      seqtype = 1</span>
<span class="c">#   Small_Diff = 1e-6</span>
<span class="c">#      verbose = 2</span>
</pre></div>
</div>
<p>However, it is usually recommended to run one model several times with different starting values, and keep only the model with higher likelihood. Here an example, on how to do this.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">best_model</span> <span class="o">=</span> <span class="bp">None</span>
<span class="n">best_lnl</span>   <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="s">&#39;-inf&#39;</span><span class="p">)</span>
<span class="k">for</span> <span class="n">starting_omega</span> <span class="ow">in</span> <span class="p">[</span><span class="mf">0.2</span><span class="p">,</span> <span class="mf">0.7</span><span class="p">,</span> <span class="mf">1.2</span><span class="p">]:</span>
    <span class="n">tree</span><span class="o">.</span><span class="n">run_model</span><span class="p">(</span><span class="s">&#39;fb.&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">starting_omega</span><span class="p">))</span>
    <span class="n">current_model</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">get_evol_model</span><span class="p">(</span><span class="s">&#39;fb.&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">starting_omega</span><span class="p">))</span>
    <span class="k">print</span> <span class="s">&#39;current model log likelihood:&#39;</span><span class="p">,</span> <span class="n">current_model</span><span class="o">.</span><span class="n">lnL</span>
    <span class="k">if</span> <span class="n">current_model</span><span class="o">.</span><span class="n">lnL</span> <span class="o">&gt;</span> <span class="n">best_lnl</span><span class="p">:</span>
        <span class="n">best_lnl</span> <span class="o">=</span> <span class="n">current_model</span><span class="o">.</span><span class="n">lnL</span>
        <span class="n">best_model</span> <span class="o">=</span> <span class="n">current_model</span>
</pre></div>
</div>
<p>Finally in order to have a quick look of the selctive pressures over our phylogeny:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">tree</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
<p>by default this will the picture obtained:</p>
<div class="figure" id="fb-default-fig">
<a class="reference internal image-reference" href="../_images/fb_default.png"><img alt="../_images/fb_default.png" src="../_images/fb_default.png" style="width: 1233.0px; height: 90.5px;" /></a>
</div>
<p>Node size, and color dependant of the <span class="math">\(\omega\)</span> value. But other displays are available:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">ete3.treeview.layouts</span> <span class="kn">import</span> <span class="n">evol_clean_layout</span>

<span class="n">tree</span><span class="o">.</span><span class="n">show</span><span class="p">(</span><span class="n">layout</span><span class="o">=</span><span class="n">evol_clean_layout</span><span class="p">)</span>
</pre></div>
</div>
<div class="figure" id="fb-clean-fig">
<a class="reference internal image-reference" href="../_images/fb_clean.png"><img alt="../_images/fb_clean.png" src="../_images/fb_clean.png" style="width: 1241.86px; height: 117.18px;" /></a>
</div>
<p>With here <span class="math">\(\omega\)</span> ratios in red and also in gray the <em>dN</em> and <em>dS</em> values.</p>
</div>
<div class="section" id="site-model">
<h3><a class="toc-backref" href="#id10">Site model</a><a class="headerlink" href="#site-model" title="Permalink to this headline">¶</a></h3>
<p>Another way to look at selective pressures, is to compute directly along the alignment, the value of <span class="math">\(\omega\)</span> for a whole column (putting all leaves together). For doing this, we can use for example the model M2 of CodeML or directly use SLR. As before we just have to:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">tree</span><span class="o">.</span><span class="n">run_model</span><span class="p">(</span><span class="s">&#39;M2&#39;</span><span class="p">)</span>
<span class="n">tree</span><span class="o">.</span><span class="n">run_model</span><span class="p">(</span><span class="s">&#39;SLR.lele&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>and to display the results:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">tree</span><span class="o">.</span><span class="n">show</span> <span class="p">(</span><span class="n">histfaces</span><span class="o">=</span><span class="p">[</span><span class="s">&#39;M2&#39;</span><span class="p">])</span>
</pre></div>
</div>
<p>when a site model is computed, an histface is automatically generated. Thus with this call, what we are doing is to draw the default histface corresponding to the model named M2.lala. This is the result:</p>
<div class="figure" id="m2-default-fig">
<a class="reference internal image-reference" href="../_images/M2_default.png"><img alt="../_images/M2_default.png" src="../_images/M2_default.png" style="width: 1572.0px; height: 261.0px;" /></a>
</div>
<p>However customizing this face is feasible:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">model2</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">get_evol_model</span> <span class="p">(</span><span class="s">&#39;M2&#39;</span><span class="p">)</span>

<span class="n">col2</span> <span class="o">=</span> <span class="p">{</span><span class="s">&#39;NS&#39;</span> <span class="p">:</span> <span class="s">&#39;black&#39;</span><span class="p">,</span> <span class="s">&#39;RX&#39;</span> <span class="p">:</span> <span class="s">&#39;black&#39;</span><span class="p">,</span>
        <span class="s">&#39;RX+&#39;</span><span class="p">:</span> <span class="s">&#39;black&#39;</span><span class="p">,</span> <span class="s">&#39;CN&#39;</span> <span class="p">:</span> <span class="s">&#39;black&#39;</span><span class="p">,</span>
        <span class="s">&#39;CN+&#39;</span><span class="p">:</span> <span class="s">&#39;black&#39;</span><span class="p">,</span> <span class="s">&#39;PS&#39;</span> <span class="p">:</span> <span class="s">&#39;black&#39;</span><span class="p">,</span> <span class="s">&#39;PS+&#39;</span><span class="p">:</span> <span class="s">&#39;black&#39;</span><span class="p">}</span>


<span class="n">model2</span><span class="o">.</span><span class="n">set_histface</span> <span class="p">(</span><span class="n">up</span><span class="o">=</span><span class="bp">False</span><span class="p">,</span> <span class="n">kind</span><span class="o">=</span><span class="s">&#39;curve&#39;</span><span class="p">,</span> <span class="n">colors</span><span class="o">=</span><span class="n">col2</span><span class="p">,</span> <span class="n">ylim</span><span class="o">=</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span><span class="mi">4</span><span class="p">],</span> <span class="n">hlines</span> <span class="o">=</span> <span class="p">[</span><span class="mf">2.5</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">,</span> <span class="mf">4.0</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">],</span> <span class="n">header</span> <span class="o">=</span> <span class="s">&#39;Many lines, error boxes, background black&#39;</span><span class="p">,</span> <span class="n">hlines_col</span><span class="o">=</span><span class="p">[</span><span class="s">&#39;orange&#39;</span><span class="p">,</span> <span class="s">&#39;yellow&#39;</span><span class="p">,</span> <span class="s">&#39;red&#39;</span><span class="p">,</span> <span class="s">&#39;cyan&#39;</span><span class="p">],</span> <span class="n">errors</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>

<span class="n">tree</span><span class="o">.</span><span class="n">show</span><span class="p">(</span><span class="n">histfaces</span><span class="o">=</span><span class="p">[</span><span class="s">&#39;M2&#39;</span><span class="p">])</span>
</pre></div>
</div>
<div class="figure" id="m2-error-fig">
<a class="reference internal image-reference" href="../_images/M2_curve.png"><img alt="../_images/M2_curve.png" src="../_images/M2_curve.png" style="width: 1572.0px; height: 261.0px;" /></a>
</div>
<p>or:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">col</span> <span class="o">=</span> <span class="p">{</span><span class="s">&#39;NS&#39;</span> <span class="p">:</span> <span class="s">&#39;grey&#39;</span><span class="p">,</span> <span class="s">&#39;RX&#39;</span> <span class="p">:</span> <span class="s">&#39;black&#39;</span><span class="p">,</span>
       <span class="s">&#39;RX+&#39;</span><span class="p">:</span> <span class="s">&#39;grey&#39;</span><span class="p">,</span> <span class="s">&#39;CN&#39;</span> <span class="p">:</span> <span class="s">&#39;black&#39;</span><span class="p">,</span>
       <span class="s">&#39;CN+&#39;</span><span class="p">:</span> <span class="s">&#39;grey&#39;</span><span class="p">,</span> <span class="s">&#39;PS&#39;</span> <span class="p">:</span> <span class="s">&#39;black&#39;</span><span class="p">,</span> <span class="s">&#39;PS+&#39;</span><span class="p">:</span> <span class="s">&#39;black&#39;</span><span class="p">}</span>
<span class="n">model2</span><span class="o">.</span><span class="n">set_histface</span> <span class="p">(</span><span class="n">up</span><span class="o">=</span><span class="bp">False</span><span class="p">,</span> <span class="n">kind</span><span class="o">=</span><span class="s">&#39;stick&#39;</span><span class="p">,</span> <span class="n">hlines</span> <span class="o">=</span> <span class="p">[</span><span class="mf">1.0</span><span class="p">,</span><span class="mf">0.3</span><span class="p">],</span> <span class="n">hlines_col</span><span class="o">=</span><span class="p">[</span><span class="s">&#39;black&#39;</span><span class="p">,</span><span class="s">&#39;grey&#39;</span><span class="p">])</span>


<span class="n">tree</span><span class="o">.</span><span class="n">show</span><span class="p">(</span><span class="n">histfaces</span><span class="o">=</span><span class="p">[</span><span class="s">&#39;M2&#39;</span><span class="p">])</span>
</pre></div>
</div>
<div class="figure" id="m2-super-profesional-fig">
<a class="reference internal image-reference" href="../_images/M2_sticks.png"><img alt="../_images/M2_sticks.png" src="../_images/M2_sticks.png" style="width: 1572.0px; height: 261.0px;" /></a>
</div>
<p>The col dictionary contains the colors for sites detected to be under positive selection (PS), relaxation (RX), or conserved (CN). However, it is not a good idea to use them now as we do not know if there is indeed positive selection.</p>
<p>To be able to accept M2 results we will have to test this model against a null model.</p>
</div>
</div>
<div class="section" id="hypothesis-testing">
<h2><a class="toc-backref" href="#id11">Hypothesis Testing</a><a class="headerlink" href="#hypothesis-testing" title="Permalink to this headline">¶</a></h2>
<p>In order to know if the parameters estimated under a given model a reliable, we have to compare its likelihood to a null model.</p>
<p>Usually, the alternative model is a model that estimates the proportion of sites with <span class="math">\(\omega &gt; 1\)</span> and we compare its likelihood with a null model, usually a model that do not (letting <span class="math">\(\omega &lt;= 1\)</span>). This comparison is done through a likelihood ratio test. If the alternative model has the best fit than we are able to accept the possibility of <span class="math">\(\omega &gt; 1\)</span>.</p>
<p>To see a non-exhaustive list of famous comparison see the documentation of the function: <a class="reference internal" href="../reference/reference_evoltree.html#ete3.EvolNode.get_most_likely" title="ete3.EvolNode.get_most_likely"><code class="xref py py-func docutils literal"><span class="pre">EvolNode.get_most_likely()</span></code></a></p>
<div class="section" id="test-on-sites">
<h3><a class="toc-backref" href="#id12">Test on sites</a><a class="headerlink" href="#test-on-sites" title="Permalink to this headline">¶</a></h3>
<p>In order to know if some sites are significantly under positive selection, relaxed or conserved we have usually to compare 2 models. However using the model &#8220;SLR&#8221; we can directly infer positive selection or relaxation through the SLR program <a class="reference internal" href="#massingham2005" id="id3">[massingham2005]</a>.</p>
<p>The most usual comparison, and perhaps the most robust, is the comparison of models M2 and M1.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">tree</span><span class="o">.</span><span class="n">run_model</span> <span class="p">(</span><span class="s">&#39;M1&#39;</span><span class="p">)</span>
<span class="n">tree</span><span class="o">.</span><span class="n">run_model</span> <span class="p">(</span><span class="s">&#39;M2&#39;</span><span class="p">)</span>

<span class="n">pval</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">get_most_likely</span> <span class="p">(</span><span class="s">&#39;M2&#39;</span><span class="p">,</span><span class="s">&#39;M1&#39;</span><span class="p">)</span>

<span class="n">model2</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">get_evol_model</span><span class="p">(</span><span class="s">&#39;M2&#39;</span><span class="p">)</span>

<span class="k">print</span> <span class="n">model2</span>

<span class="c"># Evolutionary Model M2:</span>
<span class="c">#        log likelihood       : -517.824749</span>
<span class="c">#        number of parameters : 11</span>
<span class="c">#        sites inference      : BEB, NEB</span>
<span class="c">#        sites classes        :</span>
<span class="c">#           proportions: p0=0.98794   p1=0.00000   p2=0.01206</span>
<span class="c">#           w          : w0=0.09887   w1=1.00000   w2=178.86153</span>
<span class="c">#        branches             :</span>
<span class="c">#           mark: #0, omega: 2.25526440485   , nodes paml_ids: 6 4 2 7 1 3</span>

<span class="k">if</span> <span class="n">pval</span> <span class="o">&lt;</span> <span class="mf">0.05</span><span class="p">:</span>
  <span class="k">print</span> <span class="s">&#39;M2 model wins.&#39;</span>
  <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">model2</span><span class="o">.</span><span class="n">sites</span><span class="p">[</span><span class="s">&#39;BEB&#39;</span><span class="p">][</span><span class="s">&#39;aa&#39;</span><span class="p">])):</span>
    <span class="k">if</span> <span class="n">model2</span><span class="o">.</span><span class="n">sites</span><span class="p">[</span><span class="s">&#39;BEB&#39;</span><span class="p">][</span><span class="s">&#39;p2&#39;</span><span class="p">][</span><span class="n">s</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mf">0.95</span><span class="p">:</span>
      <span class="k">print</span> <span class="s">&#39;positively selected site </span><span class="si">%s</span><span class="s"> at position: </span><span class="si">%s</span><span class="s">, with probability: </span><span class="si">%s</span><span class="s">&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">model2</span><span class="o">.</span><span class="n">sites</span><span class="p">[</span><span class="s">&#39;BEB&#39;</span><span class="p">][</span><span class="s">&#39;aa&#39;</span><span class="p">][</span><span class="n">s</span><span class="p">],</span> <span class="n">s</span><span class="o">+</span><span class="mi">1</span><span class="p">,</span> <span class="n">model2</span><span class="o">.</span><span class="n">sites</span><span class="p">[</span><span class="s">&#39;BEB&#39;</span><span class="p">][</span><span class="s">&#39;p2&#39;</span><span class="p">][</span><span class="n">s</span><span class="p">])</span>
<span class="k">else</span><span class="p">:</span>
  <span class="k">print</span> <span class="s">&#39;M1 model is not rejected&#39;</span>

<span class="c"># M2 model wins.</span>
<span class="c"># positively selected site P at position: 81, with probability: 0.96293</span>
</pre></div>
</div>
<p>Each sites model, contains a dictionary &#8216;model.sites&#8217; in which are stored the results of CodeML. These are displayed through the histface, but they are still accessible by digging a bit. &#8216;site&#8217; dictionary usually contains the result of the NEB and BEB analysis (prefer the BEB when available). For each of BEB and NEB, the probability of belonging from one category of site is summarized by &#8216;p0&#8217;, &#8216;p1&#8217; and &#8216;p2&#8217; in the case of M2 model that have only 3 class of sites (p0, the probability of belonging to the first class of sites with <span class="math">\(\omega &lt; 1\)</span>; p1, the probability of belonging to the second class of sites with <span class="math">\(\omega = 1\)</span>; p2, the probability of belonging to the third class of sites with <span class="math">\(\omega &gt; 1\)</span>). <em>For details on the models and outputs, it is recommended to read PAML documentation.</em></p>
</div>
<div class="section" id="test-on-branches">
<h3><a class="toc-backref" href="#id13">Test on branches</a><a class="headerlink" href="#test-on-branches" title="Permalink to this headline">¶</a></h3>
<p>CodeML allows to test models that infer different selective pressures on a set of given branches. To do it, some branches of the tree are marked (usually with strings like this: &#8216;#1&#8217;). This is how to do it with ETE:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">marks</span> <span class="o">=</span> <span class="p">[</span><span class="s">&#39;2&#39;</span><span class="p">,</span> <span class="s">&#39;3&#39;</span><span class="p">,</span> <span class="s">&#39;4&#39;</span><span class="p">]</span>

<span class="c"># mark a group of branches</span>
<span class="n">tree</span><span class="o">.</span><span class="n">mark_tree</span> <span class="p">(</span><span class="n">marks</span><span class="p">,</span> <span class="p">[</span><span class="s">&#39;#1&#39;</span><span class="p">,</span> <span class="s">&#39;#1&#39;</span><span class="p">,</span> <span class="s">&#39;#1&#39;</span><span class="p">])</span>
<span class="k">print</span> <span class="n">tree</span><span class="o">.</span><span class="n">write</span> <span class="p">()</span>

<span class="c"># ((Hylobates_lar,(Gorilla_gorilla #1,Pan_troglodytes #1) #1),Papio_cynocephalus);</span>
</pre></div>
</div>
<p>By doing this a branch model will compute different <span class="math">\(\omega\)</span> values in mark branches (usually called foreground <span class="math">\(\omega_{frg}\)</span>) and in the rest of the tree (usually called background <span class="math">\(\omega_{bkg}\)</span>).</p>
<dl class="docutils">
<dt>Two kinds of branch models are usually used:</dt>
<dd><ul class="first last simple">
<li>the free-branch model: were <span class="math">\(\omega_{frg}\)</span> and <span class="math">\(\omega_{bkg}\)</span> are free</li>
<li>the neutral-branch model: were <span class="math">\(\omega_{frg}\)</span> is fixed to one.</li>
<li>the M0 model: were all branches evolve at the same rate.</li>
</ul>
</dd>
</dl>
<p>The comparison between free-branch and M0, will tell us if foreground branches have an <span class="math">\(\omega\)</span> significantly different from the rest of the tree.</p>
<p>And the comparison between free-branch and neutral-branch models will tell us if <span class="math">\(\omega_{frg}\)</span> is significantly higher than 1.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">tree</span><span class="o">.</span><span class="n">run_model</span> <span class="p">(</span><span class="s">&#39;b_free&#39;</span><span class="p">)</span>
<span class="n">tree</span><span class="o">.</span><span class="n">run_model</span> <span class="p">(</span><span class="s">&#39;b_neut&#39;</span><span class="p">)</span>
<span class="n">tree</span><span class="o">.</span><span class="n">run_model</span> <span class="p">(</span><span class="s">&#39;M0&#39;</span><span class="p">)</span>

<span class="n">b_free</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">get_evol_model</span><span class="p">(</span><span class="s">&#39;b_free&#39;</span><span class="p">)</span>

<span class="k">print</span> <span class="n">b_free</span>

<span class="c"># Evolutionary Model b_free.234:</span>
<span class="c">#      log likelihood       : -525.690213</span>
<span class="c">#      number of parameters : 9</span>
<span class="c">#      sites inference      : None</span>
<span class="c">#      sites classes        : None</span>
<span class="c">#      branches             :</span>
<span class="c">#        mark: #0, omega: 0.157280451975  , nodes paml_ids: 6 4 2</span>
<span class="c">#        mark: #1, omega: 0.118462858241  , nodes paml_ids: 7 1 3</span>


<span class="k">if</span> <span class="n">tree</span><span class="o">.</span><span class="n">get_most_likely</span> <span class="p">(</span><span class="s">&#39;b_free&#39;</span><span class="p">,</span> <span class="s">&#39;M0&#39;</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mf">0.05</span><span class="p">:</span>
    <span class="c"># branch models have a branches dictionary were keys corresponds to paml_id of nodes in the tree</span>
    <span class="c"># select one of the marked branches</span>
    <span class="n">frg_node</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">search_nodes</span><span class="p">(</span><span class="n">_nid</span><span class="o">=</span><span class="mi">2</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">frg_pamlid</span> <span class="o">=</span> <span class="n">frg_node</span><span class="o">.</span><span class="n">paml_id</span>
    <span class="n">w_frg</span> <span class="o">=</span> <span class="n">bfree</span><span class="o">.</span><span class="n">branches</span><span class="p">[</span><span class="n">frg_pamlid</span><span class="p">][</span><span class="s">&#39;w&#39;</span><span class="p">]</span>
    <span class="c"># select one of the unmarked branches</span>
    <span class="n">bkg_node</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">search_nodes</span><span class="p">(</span><span class="n">_nid</span><span class="o">=</span><span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">bkg_pamlid</span> <span class="o">=</span> <span class="n">bkg_node</span><span class="o">.</span><span class="n">paml_id</span>
    <span class="n">w_bkg</span> <span class="o">=</span> <span class="n">bfree</span><span class="o">.</span><span class="n">branches</span><span class="p">[</span><span class="n">bkg_pamlid</span><span class="p">][</span><span class="s">&#39;w&#39;</span><span class="p">]</span>
    <span class="k">print</span> <span class="s">&#39;foreground branches evolving at omega value of </span><span class="si">%s</span><span class="s"> significantly diferent from </span><span class="si">%s</span><span class="s">.&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">w_frg</span><span class="p">,</span> <span class="n">w_bkg</span><span class="p">)</span>

<span class="k">if</span>  <span class="n">tree</span><span class="o">.</span><span class="n">get_most_likely</span> <span class="p">(</span><span class="s">&#39;b_free&#39;</span><span class="p">,</span> <span class="s">&#39;b_neut&#39;</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mf">0.05</span><span class="p">:</span>
    <span class="k">print</span> <span class="s">&#39;foreground branches are significantly different from 1.&#39;</span>
</pre></div>
</div>
<p><em>for the given example, background and foreground are not different...</em></p>
</div>
<div class="section" id="test-on-branch-sites">
<h3><a class="toc-backref" href="#id14">Test on branch-sites</a><a class="headerlink" href="#test-on-branch-sites" title="Permalink to this headline">¶</a></h3>
<p>An other popular test is the branch-site test contrasting model A and A1. This test allows to detect genes with some sites under positive selection.</p>
<p>this example shows how to run it over all branches in the tree:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="c"># also needs the M0 model as null model</span>
<span class="n">tree</span><span class="o">.</span><span class="n">run_model</span> <span class="p">(</span><span class="s">&#39;M0&#39;</span><span class="p">)</span>

<span class="k">for</span> <span class="n">leaf</span> <span class="ow">in</span> <span class="n">tree</span><span class="p">:</span>
    <span class="n">leaf</span><span class="o">.</span><span class="n">node_id</span>
    <span class="k">print</span> <span class="s">&#39;</span><span class="se">\n</span><span class="s">---------</span><span class="se">\n</span><span class="s">Now working with leaf &#39;</span> <span class="o">+</span> <span class="n">leaf</span><span class="o">.</span><span class="n">name</span>
    <span class="n">tree</span><span class="o">.</span><span class="n">mark_tree</span> <span class="p">([</span><span class="n">leaf</span><span class="o">.</span><span class="n">node_id</span><span class="p">],</span> <span class="n">marks</span><span class="o">=</span><span class="p">[</span><span class="s">&#39;#1&#39;</span><span class="p">])</span>
    <span class="k">print</span> <span class="n">tree</span><span class="o">.</span><span class="n">write</span><span class="p">()</span>
    <span class="c"># to organize a bit, we name model with the name of the marked node</span>
    <span class="c"># any character after the dot, in model name, is not taken into account</span>
    <span class="c"># for computation. (have a look in /tmp/ete3.../bsA.. directory)</span>
    <span class="k">print</span> <span class="s">&#39;running model bsA and bsA1&#39;</span>
    <span class="n">tree</span><span class="o">.</span><span class="n">run_model</span> <span class="p">(</span><span class="s">&#39;bsA.&#39;</span><span class="o">+</span> <span class="n">leaf</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
    <span class="n">tree</span><span class="o">.</span><span class="n">run_model</span> <span class="p">(</span><span class="s">&#39;bsA1.&#39;</span> <span class="o">+</span> <span class="n">leaf</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
    <span class="k">print</span> <span class="s">&#39;p-value of positive selection for sites on this branch is: &#39;</span>
    <span class="n">ps</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">get_most_likely</span> <span class="p">(</span><span class="s">&#39;bsA.&#39;</span> <span class="o">+</span> <span class="n">leaf</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="s">&#39;bsA1.&#39;</span><span class="o">+</span> <span class="n">leaf</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
    <span class="n">rx</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">get_most_likely</span> <span class="p">(</span><span class="s">&#39;bsA1.&#39;</span><span class="o">+</span> <span class="n">leaf</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="s">&#39;M0&#39;</span><span class="p">)</span>
    <span class="k">print</span> <span class="nb">str</span> <span class="p">(</span><span class="n">ps</span><span class="p">)</span>
    <span class="k">print</span> <span class="s">&#39;p-value of relaxation for sites on this branch is: &#39;</span>
    <span class="k">print</span> <span class="nb">str</span> <span class="p">(</span><span class="n">rx</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">ps</span><span class="o">&lt;</span><span class="mf">0.05</span> <span class="ow">and</span> <span class="nb">float</span> <span class="p">(</span><span class="n">bsA</span><span class="o">.</span><span class="n">wfrg2a</span><span class="p">)</span><span class="o">&gt;</span><span class="mi">1</span><span class="p">:</span>
        <span class="k">print</span> <span class="s">&#39;we have positive selection on sites on this branch&#39;</span>
    <span class="k">elif</span> <span class="n">rx</span><span class="o">&lt;</span><span class="mf">0.05</span> <span class="ow">and</span> <span class="n">ps</span><span class="o">&gt;=</span><span class="mf">0.05</span><span class="p">:</span>
        <span class="k">print</span> <span class="s">&#39;we have relaxation on sites on this branch&#39;</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">print</span> <span class="s">&#39;no signal detected on this branch, best fit for M0&#39;</span>
    <span class="k">print</span> <span class="s">&#39;</span><span class="se">\n</span><span class="s">clean tree, remove marks&#39;</span>
    <span class="n">tree</span><span class="o">.</span><span class="n">mark_tree</span> <span class="p">(</span><span class="nb">map</span> <span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">node_id</span><span class="p">,</span> <span class="n">tree</span><span class="o">.</span><span class="n">get_descendants</span><span class="p">()),</span>
                    <span class="n">marks</span><span class="o">=</span><span class="p">[</span><span class="s">&#39;&#39;</span><span class="p">]</span> <span class="o">*</span> <span class="nb">len</span> <span class="p">(</span><span class="n">tree</span><span class="o">.</span><span class="n">get_descendants</span><span class="p">()),</span> <span class="n">verbose</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>

<span class="k">print</span> <span class="n">tree</span><span class="o">.</span><span class="n">get_evol_model</span><span class="p">(</span><span class="s">&#39;bsA.Hylobates_lar&#39;</span><span class="p">)</span>

<span class="c">#  Evolutionary Model bsA.Hylobates_lar:</span>
<span class="c">#         log likelihood       : -521.203318</span>
<span class="c">#         number of parameters : 11</span>
<span class="c">#         sites inference      : BEB, NEB</span>
<span class="c">#         sites classes        :</span>
<span class="c">#         foreground w: f0=0.06452   f1=1.00000   f2=3.15727   f3=3.15727</span>
<span class="c">#         proportions : p0=0.00415   p1=0.00023   p2=0.94413   p3=0.05150</span>
<span class="c">#         background w: b0=0.06452   b1=1.00000   b2=0.06452   b3=1.00000</span>
<span class="c">#         branches             :</span>
<span class="c">#         mark: #0  , omega: None      , node_ids: 5   , name: NoName</span>
<span class="c">#         mark: #0  , omega: None      , node_ids: 6   , name: NoName</span>
<span class="c">#         mark: #0  , omega: None      , node_ids: 4   , name: Papio_cynocephalus</span>
<span class="c">#         mark: #1  , omega: None      , node_ids: 2   , name: Hylobates_lar</span>
<span class="c">#         mark: #0  , omega: None      , node_ids: 7   , name: NoName</span>
<span class="c">#         mark: #0  , omega: None      , node_ids: 1   , name: Gorilla_gorilla</span>
<span class="c">#         mark: #0  , omega: None      , node_ids: 3   , name: Pan_troglodytes</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="utilities">
<h2><a class="toc-backref" href="#id15">Utilities</a><a class="headerlink" href="#utilities" title="Permalink to this headline">¶</a></h2>
<div class="section" id="load-precomputed-evolutionary-model">
<h3><a class="toc-backref" href="#id16">Load precomputed evolutionary model</a><a class="headerlink" href="#load-precomputed-evolutionary-model" title="Permalink to this headline">¶</a></h3>
<p>When an evolutionary model is computed, the output is stored in <em>tree.workdir</em> and can be load afterwards. Inside <em>tree.workdir</em>, a new directory is created for each model you compute (if each model has a different name), thus to load one model:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">ete3</span> <span class="kn">import</span> <span class="n">EvolTree</span>

<span class="n">tree</span> <span class="o">=</span> <span class="n">EvolTree</span><span class="p">(</span><span class="s">&quot;((Hylobates_lar,(Gorilla_gorilla,Pan_troglodytes)),Papio_cynocephalus);&quot;</span><span class="p">)</span>

<span class="n">tree</span><span class="o">.</span><span class="n">link_to_alignment</span> <span class="p">(</span><span class="s">&#39;&#39;&#39;&gt;Hylobates_lar</span>
<span class="s">ATGGCCAGGTACAGATGCTGCCGCAGCCAGAGCCGGAGCAGATGTTACCGCCAGAGCCGGAGCAGATGTTACCGCCAGAGGCAAAGCCAGAGTCGGAGCAGATGTTACCGCCAGAGCCAGAGCCGGAGCAGATGTTACCGCCAGAGACAAAGAAGTCGGAGACGAAGGAGGCGGAGCTGCCAGACACGGAGGAGAGCCATGAGGTGT---CGCCGCAGGTACAGGCTGAGACGTAGAAGCTGTTACCACATTGTATCT</span>
<span class="s">&gt;Papio_cynocephalus</span>
<span class="s">ATGGCCAGGTACAGATGCTGCCGCAGCCAGAGCCGAAGCAGATGCTATCGCCAGAGCCGGAGCAGATGTAACCGCCAGAGACAGAGCCAAAGCCGGAGAAGCTGCTATCGCCAGAGCCAAAGCCGGAGCAGATGTTACCGCCAGAGACAGAGAAGTCGTAGACGAAGGAGGCGACGCTGCCAGACACGGAGGAGAGCCATGAGGTGCTTCCGCCGCAGGTACAGGCTGAGGCGTAGGAGGCCCTATCACATCGTGTCT</span>
<span class="s">&gt;Gorilla_gorilla</span>
<span class="s">ATGGCCAGGTACAGATGCTGTCGCAGCCAGAGCCGCAGCAGATGTTACCGGCAGAGCCGGAGCAGGTGTTACCGGCAGAGACAAAGCCAGAGCCGGAGCAGATGCTACCGGCAGAGCCAAAGCCGGAGCAGGTGTTACCGGCAGAGACAAAGAAGTCGCAGACGTAGGCGGAGGAGCTGCCAGACACGGAGGAGAGCCATGAGGTGCTGCCGCCGCAGGTACAGACTGAGACGTAGAAGACCCTATCATATTGTATCT</span>
<span class="s">&gt;Pan_troglodytes</span>
<span class="s">ATGGCCAGGTACAGATGCTGTCGCAGCCAGAGCCGGAGCAGATGTTACCGGCAGAGACGGAGCAGGTGTTACCGGCAAAGGCAAAGCCAAAGTCGGAGCAGATGTTACCGGCAGAGCCAGAGACGGAGCAGGTGTTACCGGCAAAGACAAAGAAGTCGCAGACGAAGGCGACGGAGCTGCCAGACACGGAGGAGAGCCATGAGGTGCTGCCGCCGCAGGTACAGACTGAGACGTAAAAGATGTTACCATATTGTATCT&#39;&#39;&#39;</span><span class="p">)</span>
<span class="n">tree</span><span class="o">.</span><span class="n">link_to_evol_model</span><span class="p">(</span><span class="s">&#39;/path_to/my_working_directory/fb.example/out&#39;</span><span class="p">,</span> <span class="s">&#39;fb&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p><em>Note:</em> <a class="reference internal" href="../reference/reference_evoltree.html#ete3.EvolNode.link_to_evol_model" title="ete3.EvolNode.link_to_evol_model"><code class="xref py py-func docutils literal"><span class="pre">EvolNode.link_to_evol_model()</span></code></a> is also able to load directly <code class="xref py py-class docutils literal"><span class="pre">evol.Model</span></code> objects.</p>
<p>Indeed all output files generated do not need to be kept, only a <a class="reference external" href="http://docs.python.org/library/pickle.html">pickle or cPickle</a> of the model for example.</p>
<p>Thus we can save a tree with all Models computed, after having run some models:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="c"># assume we already have run some models:</span>
<span class="k">for</span> <span class="n">model</span> <span class="ow">in</span>  <span class="n">tree</span><span class="o">.</span><span class="n">_models</span><span class="p">:</span>
    <span class="k">print</span> <span class="n">tree</span><span class="o">.</span><span class="n">get_evol_model</span><span class="p">(</span><span class="n">model</span><span class="p">)</span>

<span class="c">#  Evolutionary Model M2.example:</span>
<span class="c">#         log likelihood       : -517.824749</span>
<span class="c">#         number of parameters : 11</span>
<span class="c">#         sites inference      : BEB, NEB</span>
<span class="c">#         sites classes        :</span>
<span class="c">#            proportions : p0=0.98794   p1=0.00000   p2=0.01206</span>
<span class="c">#            w           : w0=0.09887   w1=1.00000   w2=178.86192</span>
<span class="c">#         branches             :</span>
<span class="c">#            mark: #0, omega: 2.25526864047   , nodes paml_ids: 6 4 2 7 1 3</span>
<span class="c">#</span>
<span class="c">#  Evolutionary Model M1.example:</span>
<span class="c">#         log likelihood       : -521.541845</span>
<span class="c">#         number of parameters : 9</span>
<span class="c">#         sites inference      : NEB</span>
<span class="c">#         sites classes        :</span>
<span class="c">#            proportions : p0=0.94075   p1=0.05925</span>
<span class="c">#            w           : w0=0.07025   w1=1.00000</span>
<span class="c">#         branches             :</span>
<span class="c">#            mark: #0, omega: 0.125334579074  , nodes paml_ids: 6 4 2 7 1 3</span>

<span class="c"># we save the whole tree into my_tree.pik</span>
<span class="kn">from</span> <span class="nn">cPickle</span> <span class="kn">import</span> <span class="n">dump</span>

<span class="n">out</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="s">&#39;my_tree.pik&#39;</span><span class="p">,</span> <span class="s">&#39;w&#39;</span><span class="p">)</span>
<span class="n">dump</span> <span class="p">(</span><span class="n">tree</span><span class="p">,</span> <span class="n">out</span><span class="p">)</span>
<span class="n">out</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
<p>and load them:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">cPickle</span> <span class="kn">import</span> <span class="n">load</span>


<span class="n">out</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="s">&#39;my_tree.pik&#39;</span><span class="p">)</span>
<span class="n">tree</span> <span class="o">=</span> <span class="n">load</span><span class="p">(</span><span class="n">out</span><span class="p">)</span>
<span class="n">out</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

<span class="c"># and here the same result</span>
<span class="k">for</span> <span class="n">model</span> <span class="ow">in</span>  <span class="n">tree</span><span class="o">.</span><span class="n">_models</span><span class="p">:</span>
    <span class="k">print</span> <span class="n">tree</span><span class="o">.</span><span class="n">get_evol_model</span><span class="p">(</span><span class="n">model</span><span class="p">)</span>

<span class="c">#  Evolutionary Model M2.example:</span>
<span class="c">#         log likelihood       : -517.824749</span>
<span class="c">#         number of parameters : 11</span>
<span class="c">#         sites inference      : BEB, NEB</span>
<span class="c">#         sites classes        :</span>
<span class="c">#            proportions : p0=0.98794   p1=0.00000   p2=0.01206</span>
<span class="c">#            w           : w0=0.09887   w1=1.00000   w2=178.86192</span>
<span class="c">#         branches             :</span>
<span class="c">#            mark: #0, omega: 2.25526864047   , nodes paml_ids: 6 4 2 7 1 3</span>
<span class="c">#</span>
<span class="c">#  Evolutionary Model M1.example:</span>
<span class="c">#         log likelihood       : -521.541845</span>
<span class="c">#         number of parameters : 9</span>
<span class="c">#         sites inference      : NEB</span>
<span class="c">#         sites classes        :</span>
<span class="c">#            proportions : p0=0.94075   p1=0.05925</span>
<span class="c">#            w           : w0=0.07025   w1=1.00000</span>
<span class="c">#         branches             :</span>
<span class="c">#            mark: #0, omega: 0.125334579074  , nodes paml_ids: 6 4 2 7 1 3</span>
</pre></div>
</div>
<p>this also can be done for one model alone:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">cPickle</span> <span class="kn">import</span> <span class="n">dump</span><span class="p">,</span> <span class="n">load</span>

<span class="n">m2_to_save</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">get_evol_model</span><span class="p">(</span><span class="s">&#39;M2.example&#39;</span><span class="p">)</span>

<span class="n">out</span> <span class="o">=</span> <span class="nb">open</span> <span class="p">(</span><span class="s">&#39;m2.pik&#39;</span><span class="p">,</span> <span class="s">&#39;w&#39;</span><span class="p">)</span>
<span class="n">dump</span><span class="p">(</span><span class="n">m2_to_save</span><span class="p">,</span> <span class="n">out</span><span class="p">)</span>
<span class="n">out</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

<span class="c"># and then load it</span>
<span class="n">out</span> <span class="o">=</span> <span class="nb">open</span> <span class="p">(</span><span class="s">&#39;m2.pik&#39;</span><span class="p">)</span>
<span class="n">m2_to_save</span> <span class="o">=</span> <span class="n">dump</span><span class="p">(</span><span class="n">out</span><span class="p">)</span>
<span class="n">out</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

<span class="n">tree</span><span class="o">.</span><span class="n">link_to_evol_model</span><span class="p">(</span><span class="n">m2_to_save</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="references">
<h2><a class="toc-backref" href="#id17">References</a><a class="headerlink" href="#references" title="Permalink to this headline">¶</a></h2>
<table class="docutils citation" frame="void" id="yang2007" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label"><a class="fn-backref" href="#id1">[yang2007]</a></td><td>Yang, Z., PAML 4: phylogenetic analysis by maximum likelihood. Molecular biology and evolution 24: 1586-91. (2007)</td></tr>
</tbody>
</table>
<table class="docutils citation" frame="void" id="massingham2005" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label">[massingham2005]</td><td><em>(<a class="fn-backref" href="#id2">1</a>, <a class="fn-backref" href="#id3">2</a>)</em> Massingham T. and Goldman N. Detecting amino acid sites under positive selection and purifying selection. Genetics 169: 1853-1762. (2005)</td></tr>
</tbody>
</table>
</div>
</div>


           </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="tutorial_ncbitaxonomy.html" class="btn btn-neutral float-right" title="Dealing with the NCBI Taxonomy database" accesskey="n">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="tutorial_webplugin.html" class="btn btn-neutral" title="Interactive web tree visualization" accesskey="p"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2015, The ETE Toolkit Team.
      Last updated on Mar 18, 2016.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'../',
            VERSION:'3.0.0b35',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true
        };
    </script>
      <script type="text/javascript" src="../_static/jquery.js"></script>
      <script type="text/javascript" src="../_static/underscore.js"></script>
      <script type="text/javascript" src="../_static/doctools.js"></script>
      <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>

  

  
  
    <script type="text/javascript" src="../_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>