

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Clustering Trees &mdash; ETE Toolkit - analysis and visualization of trees</title>
  

  
  
    <link rel="shortcut icon" href="../_static/favicon.ico"/>
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  

  

  
        <link rel="author" title="About these documents"
              href="../about.html"/>
    <link rel="top" title="ETE Toolkit - analysis and visualization of trees" href="../index.html"/>
        <link rel="up" title="The ETE tutorial" href="index.html"/>
        <link rel="next" title="Phylogenetic XML standards" href="tutorial_xml.html"/>
        <link rel="prev" title="Phylogenetic Trees" href="tutorial_phylogeny.html"/>
<script>
    (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
           (i[r].q=i[r].q||[]).push(arguments)
    },i[r].l=1*new Date();a=s.createElement(o),
    m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
    })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

 ga('create', 'UA-783841-4', 'etetoolkit.org');
 ga('send', 'pageview');
 </script>

<!-- 
 <script type="text/javascript">
 var _gaq = _gaq || [];
 _gaq.push(['_setAccount', 'UA-783841-12']);
 _gaq.push(['_setCookiePath', '/ete2/']);
 _gaq.push(['_trackPageview']);

 (function() {
 var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
 ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
 var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
 })();
 </script>
 -->
<!-- <link rel="canonical" href="http://etetoolkit.org/docs/2.3/CANONICALPATH" />
<meta http-equiv="refresh" content="0; url=http:/etetoolkit.org/CANONICALPATH" />
<script type="text/javascript">
window.location.replace('http://etetoolkit.org/docs/2.3/' + window.location.pathname.replace('/ete2/',''));
</script> -->
 
 


  
  <script src="../_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="../index.html" class="icon icon-home"> ETE Toolkit
          

          
          </a>

          
            
            
              <div class="version">
                3.0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
                <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../changelog/index.html">Changelog history</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">The ETE tutorial</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="tutorial_trees.html">Working With Tree Data Structures</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial_drawing.html">The Programmable Tree Drawing Engine</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial_phylogeny.html">Phylogenetic Trees</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="">Clustering Trees</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#overview">Overview</a></li>
<li class="toctree-l3"><a class="reference internal" href="#loading-clustertrees">Loading ClusterTrees</a></li>
<li class="toctree-l3"><a class="reference internal" href="#visualization-of-matrix-associated-trees">Visualization of matrix associated Trees</a></li>
<li class="toctree-l3"><a class="reference internal" href="#cluster-validation-example">Cluster Validation Example</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="tutorial_xml.html">Phylogenetic XML standards</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial_webplugin.html">Interactive web tree visualization</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial_adaptation.html">Testing Evolutionary Hypothesis</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial_ncbitaxonomy.html">Dealing with the NCBI Taxonomy database</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial_etree2orthoxml.html">SCRIPTS: orthoXML</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../reference/index.html">ETE&#8217;s Reference Guide</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="../index.html">ETE Toolkit</a>
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          





<div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="../index.html">Docs</a> &raquo;</li>
      
          <li><a href="index.html">The ETE tutorial</a> &raquo;</li>
      
    <li>Clustering Trees</li>
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="../_sources/tutorial/tutorial_clustering.txt" rel="nofollow"> View page source</a>
          
        
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <span class="target" id="module-ete3.clustering"></span><div class="section" id="clustering-trees">
<h1><a class="toc-backref" href="#id1">Clustering Trees</a><a class="headerlink" href="#clustering-trees" title="Permalink to this headline">¶</a></h1>
<div class="contents topic" id="contents">
<p class="topic-title first">Contents</p>
<ul class="simple">
<li><a class="reference internal" href="#clustering-trees" id="id1">Clustering Trees</a><ul>
<li><a class="reference internal" href="#overview" id="id2">Overview</a></li>
<li><a class="reference internal" href="#loading-clustertrees" id="id3">Loading ClusterTrees</a></li>
<li><a class="reference internal" href="#visualization-of-matrix-associated-trees" id="id4">Visualization of matrix associated Trees</a></li>
<li><a class="reference internal" href="#cluster-validation-example" id="id5">Cluster Validation Example</a></li>
</ul>
</li>
</ul>
</div>
<div class="section" id="overview">
<h2><a class="toc-backref" href="#id2">Overview</a><a class="headerlink" href="#overview" title="Permalink to this headline">¶</a></h2>
<p>Cluster analysis is the assignment of a set of observations into
subsets (called clusters) so that observations in the same cluster are
similar in some sense.  Clustering is a method of unsupervised
learning, and a common technique for statistical data analysis used in
many fields, including machine learning, data mining, pattern
recognition, image analysis and bioinformatics. Hierarchical
clustering creates a hierarchy of clusters which may be represented in
a tree structure called a dendrogram. The root of the tree consists of
a single cluster containing all observations, and the leaves
correspond to individual observations. [The Wikipedia project
Jun-2009].</p>
<p>ETE provides special <a class="reference internal" href="../reference/reference_clustering.html#ete3.ClusterNode" title="ete3.ClusterNode"><code class="xref py py-class docutils literal"><span class="pre">ClusterNode</span></code></a> (alias ClusterTree)
instances to deal with trees associated to a clustering analysis. The
basic difference between <a class="reference internal" href="../reference/reference_tree.html#ete3.Tree" title="ete3.Tree"><code class="xref py py-class docutils literal"><span class="pre">Tree</span></code></a> and <a class="reference internal" href="../reference/reference_clustering.html#ete3.ClusterTree" title="ete3.ClusterTree"><code class="xref py py-class docutils literal"><span class="pre">ClusterTree</span></code></a> is
that leaf nodes in a cluster-tree are linked to numerical
profiles. Such profiles are expected to represent the data used to
generate the clustering tree.  In other words, trees are bound to
numerical arrays.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="c">#              /-A</span>
<span class="c">#          ---|</span>
<span class="c">#              \-B</span>
<span class="c">#</span>
<span class="c"># #Names  col1   col2   col3</span>
<span class="c"># A       1.1    0.1    1.33</span>
<span class="c"># B       2.0    1.0    2.0</span>
</pre></div>
</div>
<p>Based on this, <a class="reference internal" href="../reference/reference_clustering.html#ete3.ClusterTree" title="ete3.ClusterTree"><code class="xref py py-class docutils literal"><span class="pre">ClusterTree</span></code></a> instances provide several several
clustering validation techniques that help in the analysis of cluster
quality. Currently, inter and intra-cluster distances, cluster
std.deviation, Silhouette analysis and Dunn indexes are supported. In
addition, ClusterTree nodes can be visualized using the
<a class="reference internal" href="../reference/reference_treeview.html#ete3.ProfileFace" title="ete3.ProfileFace"><code class="xref py py-class docutils literal"><span class="pre">ProfileFace</span></code></a> face type, which can represent cluster profiles
in different ways, such as line plots, heatmaps or bar plots.</p>
<p>Although this type of trees are intended to be used for clustering
results, any tree that can be linked to a table (i.e. phylogenetic
profiles) could be loaded using this data type, thus taking advantage
of the profile visualization modes, etc.</p>
</div>
<div class="section" id="loading-clustertrees">
<h2><a class="toc-backref" href="#id3">Loading ClusterTrees</a><a class="headerlink" href="#loading-clustertrees" title="Permalink to this headline">¶</a></h2>
<p>A <a class="reference internal" href="../reference/reference_clustering.html#ete3.ClusterTree" title="ete3.ClusterTree"><code class="xref py py-class docutils literal"><span class="pre">ClusterTree</span></code></a> can be linked to a numerical matrix by using
the <code class="xref py py-attr docutils literal"><span class="pre">text_array</span></code> argument.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">ete3</span> <span class="kn">import</span> <span class="n">ClusterTree</span>

<span class="c"># Example of a minimalistic numerical matrix. It is encoded as a text</span>
<span class="c"># string for convenience, but it usally be loaded from a text file.</span>
<span class="n">matrix</span> <span class="o">=</span> <span class="s">&quot;&quot;&quot;</span>
<span class="s">#Names</span><span class="se">\t</span><span class="s">col1</span><span class="se">\t</span><span class="s">col2</span><span class="se">\t</span><span class="s">col3</span><span class="se">\t</span><span class="s">col4</span><span class="se">\t</span><span class="s">col5</span><span class="se">\t</span><span class="s">col6</span><span class="se">\t</span><span class="s">col7</span>
<span class="s">A</span><span class="se">\t</span><span class="s">-1.23</span><span class="se">\t</span><span class="s">-0.81</span><span class="se">\t</span><span class="s">1.79</span><span class="se">\t</span><span class="s">0.78</span><span class="se">\t</span><span class="s">-0.42</span><span class="se">\t</span><span class="s">-0.69</span><span class="se">\t</span><span class="s">0.58</span>
<span class="s">B</span><span class="se">\t</span><span class="s">-1.76</span><span class="se">\t</span><span class="s">-0.94</span><span class="se">\t</span><span class="s">1.16</span><span class="se">\t</span><span class="s">0.36</span><span class="se">\t</span><span class="s">0.41</span><span class="se">\t</span><span class="s">-0.35</span><span class="se">\t</span><span class="s">1.12</span>
<span class="s">C</span><span class="se">\t</span><span class="s">-2.19</span><span class="se">\t</span><span class="s">0.13</span><span class="se">\t</span><span class="s">0.65</span><span class="se">\t</span><span class="s">-0.51</span><span class="se">\t</span><span class="s">0.52</span><span class="se">\t</span><span class="s">1.04</span><span class="se">\t</span><span class="s">0.36</span>
<span class="s">D</span><span class="se">\t</span><span class="s">-1.22</span><span class="se">\t</span><span class="s">-0.98</span><span class="se">\t</span><span class="s">0.79</span><span class="se">\t</span><span class="s">-0.76</span><span class="se">\t</span><span class="s">-0.29</span><span class="se">\t</span><span class="s">1.54</span><span class="se">\t</span><span class="s">0.93</span>
<span class="s">E</span><span class="se">\t</span><span class="s">-1.47</span><span class="se">\t</span><span class="s">-0.83</span><span class="se">\t</span><span class="s">0.85</span><span class="se">\t</span><span class="s">0.07</span><span class="se">\t</span><span class="s">-0.81</span><span class="se">\t</span><span class="s">1.53</span><span class="se">\t</span><span class="s">0.65</span>
<span class="s">F</span><span class="se">\t</span><span class="s">-1.04</span><span class="se">\t</span><span class="s">-1.11</span><span class="se">\t</span><span class="s">0.87</span><span class="se">\t</span><span class="s">-0.14</span><span class="se">\t</span><span class="s">-0.80</span><span class="se">\t</span><span class="s">1.74</span><span class="se">\t</span><span class="s">0.48</span>
<span class="s">G</span><span class="se">\t</span><span class="s">-1.57</span><span class="se">\t</span><span class="s">-1.17</span><span class="se">\t</span><span class="s">1.29</span><span class="se">\t</span><span class="s">0.23</span><span class="se">\t</span><span class="s">-0.20</span><span class="se">\t</span><span class="s">1.17</span><span class="se">\t</span><span class="s">0.26</span>
<span class="s">H</span><span class="se">\t</span><span class="s">-1.53</span><span class="se">\t</span><span class="s">-1.25</span><span class="se">\t</span><span class="s">0.59</span><span class="se">\t</span><span class="s">-0.30</span><span class="se">\t</span><span class="s">0.32</span><span class="se">\t</span><span class="s">1.41</span><span class="se">\t</span><span class="s">0.77</span>
<span class="s">&quot;&quot;&quot;</span>
<span class="k">print</span> <span class="s">&quot;Example numerical matrix&quot;</span>
<span class="k">print</span> <span class="n">matrix</span>
<span class="c"># #Names  col1    col2    col3    col4    col5    col6    col7</span>
<span class="c"># A       -1.23   -0.81   1.79    0.78    -0.42   -0.69   0.58</span>
<span class="c"># B       -1.76   -0.94   1.16    0.36    0.41    -0.35   1.12</span>
<span class="c"># C       -2.19   0.13    0.65    -0.51   0.52    1.04    0.36</span>
<span class="c"># D       -1.22   -0.98   0.79    -0.76   -0.29   1.54    0.93</span>
<span class="c"># E       -1.47   -0.83   0.85    0.07    -0.81   1.53    0.65</span>
<span class="c"># F       -1.04   -1.11   0.87    -0.14   -0.80   1.74    0.48</span>
<span class="c"># G       -1.57   -1.17   1.29    0.23    -0.20   1.17    0.26</span>
<span class="c"># H       -1.53   -1.25   0.59    -0.30   0.32    1.41    0.77</span>
<span class="c">#</span>
<span class="c">#</span>
<span class="c"># We load a tree structure whose leaf nodes correspond to rows in the</span>
<span class="c"># numerical matrix. We use the text_array argument to link the tree</span>
<span class="c"># with numerical matrix.</span>
<span class="n">t</span> <span class="o">=</span> <span class="n">ClusterTree</span><span class="p">(</span><span class="s">&quot;(((A,B),(C,(D,E))),(F,(G,H)));&quot;</span><span class="p">,</span> <span class="n">text_array</span><span class="o">=</span><span class="n">matrix</span><span class="p">)</span>
</pre></div>
</div>
<p>Alternatively, you can re-link the tree (or a sub-part of it) to a new
matrix using the <a class="reference internal" href="../reference/reference_clustering.html#ete3.ClusterNode.link_to_arraytable" title="ete3.ClusterNode.link_to_arraytable"><code class="xref py py-func docutils literal"><span class="pre">ClusterNode.link_to_arraytable()</span></code></a> method.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">t</span> <span class="o">=</span> <span class="n">ClusterTree</span><span class="p">(</span><span class="s">&quot;(((A,B),(C,(D,E))),(F,(G,H)));&quot;</span><span class="p">)</span>
<span class="n">t</span><span class="o">.</span><span class="n">children</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">link_to_arraytable</span><span class="p">(</span><span class="n">matrix1</span><span class="p">)</span>
<span class="n">t</span><span class="o">.</span><span class="n">children</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">link_to_arraytable</span><span class="p">(</span><span class="n">matrix2</span><span class="p">)</span>
</pre></div>
</div>
<p>Once the tree is linked to a table of profiles, the following node
properties will be available: <code class="xref py py-attr docutils literal"><span class="pre">PhyloNode.profile</span></code>,
<code class="xref py py-attr docutils literal"><span class="pre">PhyloNode.deviation</span></code>, <code class="xref py py-attr docutils literal"><span class="pre">PhyloNode.silhoutte</span></code>,
<code class="xref py py-attr docutils literal"><span class="pre">PhyloNode.intercluster_dist</span></code>,
<code class="xref py py-attr docutils literal"><span class="pre">PhyloNode.intracluster_dist</span></code>, <code class="xref py py-attr docutils literal"><span class="pre">PhyloNode.dunn</span></code>.</p>
<p>Similarly, the following methods are provide for convenience
<code class="xref py py-func docutils literal"><span class="pre">PhyloNode.iter_leaf_profiles()</span></code>,
<code class="xref py py-func docutils literal"><span class="pre">PhyloNode.get_leaf_profiles()</span></code>, <code class="xref py py-func docutils literal"><span class="pre">PhyloNode.get_silhouette()</span></code>
and <code class="xref py py-func docutils literal"><span class="pre">PhyloNode.get_dunn()</span></code> methods.</p>
</div>
<div class="section" id="visualization-of-matrix-associated-trees">
<h2><a class="toc-backref" href="#id4">Visualization of matrix associated Trees</a><a class="headerlink" href="#visualization-of-matrix-associated-trees" title="Permalink to this headline">¶</a></h2>
<p>Clustering or not, any ClusterTree instance, associated to a numerical
matrix, can be visualized together with the graphical representation
of its node&#8217;s numeric profiles. To this end, the <a class="reference internal" href="../reference/reference_treeview.html#ete3.ProfileFace" title="ete3.ProfileFace"><code class="xref py py-class docutils literal"><span class="pre">ProfileFace</span></code></a>
class is provided by the <code class="xref py py-mod docutils literal"><span class="pre">treeview</span></code> module. This face type can
represent a node&#8217;s numeric profile in four different ways:</p>
<blockquote>
<div><div class="figure">
<img alt="../_images/profile_faces.png" src="../_images/profile_faces.png" />
</div>
</div></blockquote>
<p>Additionally, three basic layouts are provided that use different
styles of ProfileFace instances: <strong>heatmap</strong>, <strong>line_profiles</strong>,
<strong>bar_profiles</strong>, <strong>cbar_profiles</strong>.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="c"># Import Tree instance and faces module</span>
<span class="kn">from</span> <span class="nn">ete3</span> <span class="kn">import</span> <span class="n">ClusterTree</span>
<span class="c"># Example of a minimalistic numerical matrix. It is encoded as a text</span>
<span class="c"># string for convenience, but it usally be loaded from a text file.</span>
<span class="n">matrix</span> <span class="o">=</span> <span class="s">&quot;&quot;&quot;</span>
<span class="s">#Names</span><span class="se">\t</span><span class="s">col1</span><span class="se">\t</span><span class="s">col2</span><span class="se">\t</span><span class="s">col3</span><span class="se">\t</span><span class="s">col4</span><span class="se">\t</span><span class="s">col5</span><span class="se">\t</span><span class="s">col6</span><span class="se">\t</span><span class="s">col7</span>
<span class="s">A</span><span class="se">\t</span><span class="s">-1.23</span><span class="se">\t</span><span class="s">-0.81</span><span class="se">\t</span><span class="s">1.79</span><span class="se">\t</span><span class="s">0.78</span><span class="se">\t</span><span class="s">-0.42</span><span class="se">\t</span><span class="s">-0.69</span><span class="se">\t</span><span class="s">0.58</span>
<span class="s">B</span><span class="se">\t</span><span class="s">-1.76</span><span class="se">\t</span><span class="s">-0.94</span><span class="se">\t</span><span class="s">1.16</span><span class="se">\t</span><span class="s">0.36</span><span class="se">\t</span><span class="s">0.41</span><span class="se">\t</span><span class="s">-0.35</span><span class="se">\t</span><span class="s">1.12</span>
<span class="s">C</span><span class="se">\t</span><span class="s">-2.19</span><span class="se">\t</span><span class="s">0.13</span><span class="se">\t</span><span class="s">0.65</span><span class="se">\t</span><span class="s">-0.51</span><span class="se">\t</span><span class="s">0.52</span><span class="se">\t</span><span class="s">1.04</span><span class="se">\t</span><span class="s">0.36</span>
<span class="s">D</span><span class="se">\t</span><span class="s">-1.22</span><span class="se">\t</span><span class="s">-0.98</span><span class="se">\t</span><span class="s">0.79</span><span class="se">\t</span><span class="s">-0.76</span><span class="se">\t</span><span class="s">-0.29</span><span class="se">\t</span><span class="s">1.54</span><span class="se">\t</span><span class="s">0.93</span>
<span class="s">E</span><span class="se">\t</span><span class="s">-1.47</span><span class="se">\t</span><span class="s">-0.83</span><span class="se">\t</span><span class="s">0.85</span><span class="se">\t</span><span class="s">0.07</span><span class="se">\t</span><span class="s">-0.81</span><span class="se">\t</span><span class="s">1.53</span><span class="se">\t</span><span class="s">0.65</span>
<span class="s">F</span><span class="se">\t</span><span class="s">-1.04</span><span class="se">\t</span><span class="s">-1.11</span><span class="se">\t</span><span class="s">0.87</span><span class="se">\t</span><span class="s">-0.14</span><span class="se">\t</span><span class="s">-0.80</span><span class="se">\t</span><span class="s">1.74</span><span class="se">\t</span><span class="s">0.48</span>
<span class="s">G</span><span class="se">\t</span><span class="s">-1.57</span><span class="se">\t</span><span class="s">-1.17</span><span class="se">\t</span><span class="s">1.29</span><span class="se">\t</span><span class="s">0.23</span><span class="se">\t</span><span class="s">-0.20</span><span class="se">\t</span><span class="s">1.17</span><span class="se">\t</span><span class="s">0.26</span>
<span class="s">H</span><span class="se">\t</span><span class="s">-1.53</span><span class="se">\t</span><span class="s">-1.25</span><span class="se">\t</span><span class="s">0.59</span><span class="se">\t</span><span class="s">-0.30</span><span class="se">\t</span><span class="s">0.32</span><span class="se">\t</span><span class="s">1.41</span><span class="se">\t</span><span class="s">0.77</span>
<span class="s">&quot;&quot;&quot;</span>
<span class="k">print</span> <span class="s">&quot;Example numerical matrix&quot;</span>
<span class="k">print</span> <span class="n">matrix</span>
<span class="c"># #Names  col1    col2    col3    col4    col5    col6    col7</span>
<span class="c"># A       -1.23   -0.81   1.79    0.78    -0.42   -0.69   0.58</span>
<span class="c"># B       -1.76   -0.94   1.16    0.36    0.41    -0.35   1.12</span>
<span class="c"># C       -2.19   0.13    0.65    -0.51   0.52    1.04    0.36</span>
<span class="c"># D       -1.22   -0.98   0.79    -0.76   -0.29   1.54    0.93</span>
<span class="c"># E       -1.47   -0.83   0.85    0.07    -0.81   1.53    0.65</span>
<span class="c"># F       -1.04   -1.11   0.87    -0.14   -0.80   1.74    0.48</span>
<span class="c"># G       -1.57   -1.17   1.29    0.23    -0.20   1.17    0.26</span>
<span class="c"># H       -1.53   -1.25   0.59    -0.30   0.32    1.41    0.77</span>
<span class="c">#</span>
<span class="c">#</span>
<span class="c"># We load a tree structure whose leaf nodes correspond to rows in the</span>
<span class="c"># numerical matrix. We use the text_array argument to link the tree</span>
<span class="c"># with numerical matrix.</span>
<span class="n">t</span> <span class="o">=</span> <span class="n">ClusterTree</span><span class="p">(</span><span class="s">&quot;(((A,B),(C,(D,E))),(F,(G,H)));&quot;</span><span class="p">,</span> <span class="n">text_array</span><span class="o">=</span><span class="n">matrix</span><span class="p">)</span>
<span class="n">t</span><span class="o">.</span><span class="n">show</span><span class="p">(</span><span class="s">&quot;heatmap&quot;</span><span class="p">)</span>
<span class="n">t</span><span class="o">.</span><span class="n">show</span><span class="p">(</span><span class="s">&quot;cluster_cbars&quot;</span><span class="p">)</span>
<span class="n">t</span><span class="o">.</span><span class="n">show</span><span class="p">(</span><span class="s">&quot;cluster_bars&quot;</span><span class="p">)</span>
<span class="n">t</span><span class="o">.</span><span class="n">show</span><span class="p">(</span><span class="s">&quot;cluster_lines&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="cluster-validation-example">
<h2><a class="toc-backref" href="#id5">Cluster Validation Example</a><a class="headerlink" href="#cluster-validation-example" title="Permalink to this headline">¶</a></h2>
<p>If associated matrix represents the dataset used to produce a given
tree, clustering validation values can be used to assess the quality
of partitions. To do so, you will need to set the distance function
that was used to calculate distances among items (leaf nodes). ETE
implements three common distance methods in bioinformatics :
<strong>euclidean</strong>, <strong>pearson</strong> correlation and <strong>spearman</strong> correlation
distances.</p>
<p>In the following example, a microarray clustering result is visualized
and validated using ETE.</p>
<div class="figure">
<img alt="../_images/bubbles_validation.png" src="../_images/bubbles_validation.png" />
</div>
<p>Image resulting from a microarray clustering validation analysis. Red
bubbles represent a bad silhouette index (S&lt;0), while green represents
good silhouette index (S&gt;0). Size of bubbles is proportional to the
Silhouette index. Internal nodes are drawn with the average expression
profile grouped by their partitions. Leaf node profiles are shown as a
heatmap.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">ete3</span> <span class="kn">import</span> <span class="n">ClusterTree</span><span class="p">,</span> <span class="n">TreeStyle</span><span class="p">,</span> <span class="n">AttrFace</span><span class="p">,</span> <span class="n">ProfileFace</span><span class="p">,</span> <span class="n">TextFace</span>
<span class="kn">from</span> <span class="nn">ete3.treeview.faces</span> <span class="kn">import</span> <span class="n">add_face_to_node</span>

<span class="c"># To operate with numbers efficiently</span>
<span class="kn">import</span> <span class="nn">numpy</span>

<span class="n">PATH</span> <span class="o">=</span> <span class="s">&quot;./&quot;</span>
<span class="c"># Loads tree and array</span>
<span class="n">t</span> <span class="o">=</span> <span class="n">ClusterTree</span><span class="p">(</span><span class="n">PATH</span><span class="o">+</span><span class="s">&quot;diauxic.nw&quot;</span><span class="p">,</span> <span class="n">PATH</span><span class="o">+</span><span class="s">&quot;diauxic.array&quot;</span><span class="p">)</span>

<span class="c"># nodes are linked to the array table</span>
<span class="n">array</span> <span class="o">=</span>  <span class="n">t</span><span class="o">.</span><span class="n">arraytable</span>

<span class="c"># Calculates some stats on the matrix. Needed to establish the color</span>
<span class="c"># gradients.</span>
<span class="n">matrix_dist</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span> <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="nb">xrange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">array</span><span class="o">.</span><span class="n">matrix</span><span class="p">))</span>\
               <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">array</span><span class="o">.</span><span class="n">matrix</span><span class="p">[</span><span class="n">r</span><span class="p">]</span> <span class="k">if</span> <span class="n">numpy</span><span class="o">.</span><span class="n">isfinite</span><span class="p">(</span><span class="n">i</span><span class="p">)]</span>
<span class="n">matrix_max</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">matrix_dist</span><span class="p">)</span>
<span class="n">matrix_min</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">min</span><span class="p">(</span><span class="n">matrix_dist</span><span class="p">)</span>
<span class="n">matrix_avg</span> <span class="o">=</span> <span class="n">matrix_min</span><span class="o">+</span><span class="p">((</span><span class="n">matrix_max</span><span class="o">-</span><span class="n">matrix_min</span><span class="p">)</span><span class="o">/</span><span class="mi">2</span><span class="p">)</span>

<span class="c"># Creates a profile face that will represent node&#39;s profile as a</span>
<span class="c"># heatmap</span>
<span class="n">profileFace</span>  <span class="o">=</span> <span class="n">ProfileFace</span><span class="p">(</span><span class="n">matrix_max</span><span class="p">,</span> <span class="n">matrix_min</span><span class="p">,</span> <span class="n">matrix_avg</span><span class="p">,</span> \
                                         <span class="mi">200</span><span class="p">,</span> <span class="mi">14</span><span class="p">,</span> <span class="s">&quot;heatmap&quot;</span><span class="p">)</span>
<span class="n">cbarsFace</span> <span class="o">=</span> <span class="n">ProfileFace</span><span class="p">(</span><span class="n">matrix_max</span><span class="p">,</span><span class="n">matrix_min</span><span class="p">,</span><span class="n">matrix_avg</span><span class="p">,</span><span class="mi">200</span><span class="p">,</span><span class="mi">70</span><span class="p">,</span><span class="s">&quot;cbars&quot;</span><span class="p">)</span>
<span class="n">nameFace</span> <span class="o">=</span> <span class="n">AttrFace</span><span class="p">(</span><span class="s">&quot;name&quot;</span><span class="p">,</span> <span class="n">fsize</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span>
<span class="c"># Creates my own layout function that uses previous faces</span>
<span class="k">def</span> <span class="nf">mylayout</span><span class="p">(</span><span class="n">node</span><span class="p">):</span>
    <span class="c"># If node is a leaf</span>
    <span class="k">if</span> <span class="n">node</span><span class="o">.</span><span class="n">is_leaf</span><span class="p">():</span>
        <span class="c"># And a line profile</span>
        <span class="n">add_face_to_node</span><span class="p">(</span><span class="n">profileFace</span><span class="p">,</span> <span class="n">node</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="n">aligned</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
        <span class="n">node</span><span class="o">.</span><span class="n">img_style</span><span class="p">[</span><span class="s">&quot;size&quot;</span><span class="p">]</span><span class="o">=</span><span class="mi">0</span>
        <span class="n">add_face_to_node</span><span class="p">(</span><span class="n">nameFace</span><span class="p">,</span> <span class="n">node</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">aligned</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>

    <span class="c"># If node is internal</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="c"># If silhouette is good, creates a green bubble</span>
        <span class="k">if</span> <span class="n">node</span><span class="o">.</span><span class="n">silhouette</span><span class="o">&gt;</span><span class="mi">0</span><span class="p">:</span>
            <span class="n">validationFace</span> <span class="o">=</span> <span class="n">TextFace</span><span class="p">(</span><span class="s">&quot;Silh=</span><span class="si">%0.2f</span><span class="s">&quot;</span> <span class="o">%</span><span class="n">node</span><span class="o">.</span><span class="n">silhouette</span><span class="p">,</span>
                                      <span class="s">&quot;Verdana&quot;</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="s">&quot;#056600&quot;</span><span class="p">)</span>
            <span class="n">node</span><span class="o">.</span><span class="n">img_style</span><span class="p">[</span><span class="s">&quot;fgcolor&quot;</span><span class="p">]</span><span class="o">=</span><span class="s">&quot;#056600&quot;</span>
        <span class="c"># Otherwise, use red bubbles</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">validationFace</span> <span class="o">=</span> <span class="n">TextFace</span><span class="p">(</span><span class="s">&quot;Silh=</span><span class="si">%0.2f</span><span class="s">&quot;</span> <span class="o">%</span><span class="n">node</span><span class="o">.</span><span class="n">silhouette</span><span class="p">,</span>
                                      <span class="s">&quot;Verdana&quot;</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="s">&quot;#940000&quot;</span><span class="p">)</span>
            <span class="n">node</span><span class="o">.</span><span class="n">img_style</span><span class="p">[</span><span class="s">&quot;fgcolor&quot;</span><span class="p">]</span><span class="o">=</span><span class="s">&quot;#940000&quot;</span>

        <span class="c"># Sets node size proportional to the silhouette value.</span>
        <span class="n">node</span><span class="o">.</span><span class="n">img_style</span><span class="p">[</span><span class="s">&quot;shape&quot;</span><span class="p">]</span><span class="o">=</span><span class="s">&quot;sphere&quot;</span>
        <span class="k">if</span> <span class="n">node</span><span class="o">.</span><span class="n">silhouette</span><span class="o">&lt;=</span><span class="mi">1</span> <span class="ow">and</span> <span class="n">node</span><span class="o">.</span><span class="n">silhouette</span><span class="o">&gt;=-</span><span class="mi">1</span><span class="p">:</span>
            <span class="n">node</span><span class="o">.</span><span class="n">img_style</span><span class="p">[</span><span class="s">&quot;size&quot;</span><span class="p">]</span><span class="o">=</span> <span class="mi">15</span><span class="o">+</span><span class="nb">int</span><span class="p">((</span><span class="nb">abs</span><span class="p">(</span><span class="n">node</span><span class="o">.</span><span class="n">silhouette</span><span class="p">)</span><span class="o">*</span><span class="mi">10</span><span class="p">)</span><span class="o">**</span><span class="mi">2</span><span class="p">)</span>

        <span class="c"># If node is very internal, draw also a bar diagram</span>
        <span class="c"># with the average expression of the partition</span>
        <span class="n">add_face_to_node</span><span class="p">(</span><span class="n">validationFace</span><span class="p">,</span> <span class="n">node</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">node</span><span class="p">)</span><span class="o">&gt;</span><span class="mi">100</span><span class="p">:</span>
            <span class="n">add_face_to_node</span><span class="p">(</span><span class="n">cbarsFace</span><span class="p">,</span> <span class="n">node</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>

<span class="c"># Use my layout to visualize the tree</span>
<span class="n">ts</span> <span class="o">=</span> <span class="n">TreeStyle</span><span class="p">()</span>
<span class="n">ts</span><span class="o">.</span><span class="n">layout_fn</span> <span class="o">=</span> <span class="n">mylayout</span>
<span class="n">t</span><span class="o">.</span><span class="n">show</span><span class="p">(</span><span class="n">tree_style</span><span class="o">=</span><span class="n">ts</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>


           </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="tutorial_xml.html" class="btn btn-neutral float-right" title="Phylogenetic XML standards" accesskey="n">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="tutorial_phylogeny.html" class="btn btn-neutral" title="Phylogenetic Trees" accesskey="p"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2015, The ETE Toolkit Team.
      Last updated on Mar 18, 2016.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'../',
            VERSION:'3.0.0b35',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true
        };
    </script>
      <script type="text/javascript" src="../_static/jquery.js"></script>
      <script type="text/javascript" src="../_static/underscore.js"></script>
      <script type="text/javascript" src="../_static/doctools.js"></script>
      <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>

  

  
  
    <script type="text/javascript" src="../_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>