

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Phylogenetic Trees &mdash; ETE Toolkit - analysis and visualization of trees</title>
  

  
  
    <link rel="shortcut icon" href="../_static/favicon.ico"/>
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  

  

  
        <link rel="author" title="About these documents"
              href="../about.html"/>
    <link rel="top" title="ETE Toolkit - analysis and visualization of trees" href="../index.html"/>
        <link rel="up" title="The ETE tutorial" href="index.html"/>
        <link rel="next" title="Clustering Trees" href="tutorial_clustering.html"/>
        <link rel="prev" title="The Programmable Tree Drawing Engine" href="tutorial_drawing.html"/>
<script>
    (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
           (i[r].q=i[r].q||[]).push(arguments)
    },i[r].l=1*new Date();a=s.createElement(o),
    m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
    })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

 ga('create', 'UA-783841-4', 'etetoolkit.org');
 ga('send', 'pageview');
 </script>

<!-- 
 <script type="text/javascript">
 var _gaq = _gaq || [];
 _gaq.push(['_setAccount', 'UA-783841-12']);
 _gaq.push(['_setCookiePath', '/ete2/']);
 _gaq.push(['_trackPageview']);

 (function() {
 var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
 ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
 var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
 })();
 </script>
 -->
<!-- <link rel="canonical" href="http://etetoolkit.org/docs/2.3/CANONICALPATH" />
<meta http-equiv="refresh" content="0; url=http:/etetoolkit.org/CANONICALPATH" />
<script type="text/javascript">
window.location.replace('http://etetoolkit.org/docs/2.3/' + window.location.pathname.replace('/ete2/',''));
</script> -->
 
 


  
  <script src="../_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="../index.html" class="icon icon-home"> ETE Toolkit
          

          
          </a>

          
            
            
              <div class="version">
                3.0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
                <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../changelog/index.html">Changelog history</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">The ETE tutorial</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="tutorial_trees.html">Working With Tree Data Structures</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial_drawing.html">The Programmable Tree Drawing Engine</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="">Phylogenetic Trees</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#overview">Overview</a></li>
<li class="toctree-l3"><a class="reference internal" href="#linking-phylogenetic-trees-with-multiple-sequence-alignments">Linking Phylogenetic Trees with Multiple Sequence Alignments</a></li>
<li class="toctree-l3"><a class="reference internal" href="#visualization-of-phylogenetic-trees">Visualization of phylogenetic trees</a></li>
<li class="toctree-l3"><a class="reference internal" href="#adding-taxonomic-information">Adding taxonomic information</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#automatic-control-of-species-info">Automatic control of species info</a></li>
<li class="toctree-l4"><a class="reference internal" href="#automatic-and-custom-control-of-the-species-info">Automatic (and custom) control of the species info</a></li>
<li class="toctree-l4"><a class="reference internal" href="#manual-control-of-the-species-info">Manual control of the species info</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#detecting-evolutionary-events">Detecting evolutionary events</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#species-overlap-so-algorithm">Species Overlap (SO) algorithm</a></li>
<li class="toctree-l4"><a class="reference internal" href="#tree-reconciliation-algorithm">Tree reconciliation algorithm</a></li>
<li class="toctree-l4"><a class="reference internal" href="#a-closer-look-to-the-evolutionary-event-object">A closer look to the evolutionary event object</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#relative-dating-phylogenetic-nodes">Relative dating phylogenetic nodes</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#implementation">Implementation</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#automatic-rooting-outgroup-detection">Automatic rooting (outgroup detection)</a></li>
<li class="toctree-l3"><a class="reference internal" href="#working-with-duplicated-gene-families">Working with duplicated gene families</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#treeko-splitting-gene-trees-into-species-trees">Treeko (splitting gene trees into species trees)</a></li>
<li class="toctree-l4"><a class="reference internal" href="#splitting-gene-trees-by-duplication-events">Splitting gene trees by duplication events</a></li>
<li class="toctree-l4"><a class="reference internal" href="#collapse-species-specific-duplications">Collapse species specific duplications</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="tutorial_clustering.html">Clustering Trees</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial_xml.html">Phylogenetic XML standards</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial_webplugin.html">Interactive web tree visualization</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial_adaptation.html">Testing Evolutionary Hypothesis</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial_ncbitaxonomy.html">Dealing with the NCBI Taxonomy database</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial_etree2orthoxml.html">SCRIPTS: orthoXML</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../reference/index.html">ETE&#8217;s Reference Guide</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="../index.html">ETE Toolkit</a>
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          





<div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="../index.html">Docs</a> &raquo;</li>
      
          <li><a href="index.html">The ETE tutorial</a> &raquo;</li>
      
    <li>Phylogenetic Trees</li>
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="../_sources/tutorial/tutorial_phylogeny.txt" rel="nofollow"> View page source</a>
          
        
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="phylogenetic-trees">
<h1><a class="toc-backref" href="#id1">Phylogenetic Trees</a><a class="headerlink" href="#phylogenetic-trees" title="Permalink to this headline">Â¶</a></h1>
<div class="contents topic" id="contents">
<p class="topic-title first">Contents</p>
<ul class="simple">
<li><a class="reference internal" href="#phylogenetic-trees" id="id1">Phylogenetic Trees</a><ul>
<li><a class="reference internal" href="#overview" id="id2">Overview</a></li>
<li><a class="reference internal" href="#linking-phylogenetic-trees-with-multiple-sequence-alignments" id="id3">Linking Phylogenetic Trees with Multiple Sequence Alignments</a></li>
<li><a class="reference internal" href="#visualization-of-phylogenetic-trees" id="id4">Visualization of phylogenetic trees</a></li>
<li><a class="reference internal" href="#adding-taxonomic-information" id="id5">Adding taxonomic information</a><ul>
<li><a class="reference internal" href="#automatic-control-of-species-info" id="id6">Automatic control of species info</a></li>
<li><a class="reference internal" href="#automatic-and-custom-control-of-the-species-info" id="id7">Automatic (and custom) control of the species info</a></li>
<li><a class="reference internal" href="#manual-control-of-the-species-info" id="id8">Manual control of the species info</a></li>
</ul>
</li>
<li><a class="reference internal" href="#detecting-evolutionary-events" id="id9">Detecting evolutionary events</a><ul>
<li><a class="reference internal" href="#species-overlap-so-algorithm" id="id10">Species Overlap (SO) algorithm</a></li>
<li><a class="reference internal" href="#tree-reconciliation-algorithm" id="id11">Tree reconciliation algorithm</a></li>
<li><a class="reference internal" href="#a-closer-look-to-the-evolutionary-event-object" id="id12">A closer look to the evolutionary event object</a></li>
</ul>
</li>
<li><a class="reference internal" href="#relative-dating-phylogenetic-nodes" id="id13">Relative dating phylogenetic nodes</a><ul>
<li><a class="reference internal" href="#implementation" id="id14">Implementation</a></li>
</ul>
</li>
<li><a class="reference internal" href="#automatic-rooting-outgroup-detection" id="id15">Automatic rooting (outgroup detection)</a></li>
<li><a class="reference internal" href="#working-with-duplicated-gene-families" id="id16">Working with duplicated gene families</a><ul>
<li><a class="reference internal" href="#treeko-splitting-gene-trees-into-species-trees" id="id17">Treeko (splitting gene trees into species trees)</a></li>
<li><a class="reference internal" href="#splitting-gene-trees-by-duplication-events" id="id18">Splitting gene trees by duplication events</a></li>
<li><a class="reference internal" href="#collapse-species-specific-duplications" id="id19">Collapse species specific duplications</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
<div class="section" id="overview">
<h2><a class="toc-backref" href="#id2">Overview</a><a class="headerlink" href="#overview" title="Permalink to this headline">Â¶</a></h2>
<p>Phylogenetic trees are the result of most evolutionary analyses. They
represent the evolutionary relationships among a set of species or, in
molecular biology, a set of homologous sequences.</p>
<p>The <a class="reference internal" href="../reference/reference_phylo.html#ete3.PhyloTree" title="ete3.PhyloTree"><code class="xref py py-class docutils literal"><span class="pre">PhyloTree</span></code></a> class is an extension of the base <a class="reference internal" href="../reference/reference_tree.html#ete3.Tree" title="ete3.Tree"><code class="xref py py-class docutils literal"><span class="pre">Tree</span></code></a>
object, providing a appropriate way to deal with phylogenetic trees.
Thus, while leaves are considered to represent species (or sequences
from a given species genome), internal nodes are considered ancestral
nodes. A direct consequence of this is, for instance, that every split
in the tree will represent a speciation or duplication event.</p>
</div>
<div class="section" id="linking-phylogenetic-trees-with-multiple-sequence-alignments">
<h2><a class="toc-backref" href="#id3">Linking Phylogenetic Trees with Multiple Sequence Alignments</a><a class="headerlink" href="#linking-phylogenetic-trees-with-multiple-sequence-alignments" title="Permalink to this headline">Â¶</a></h2>
<p><a class="reference internal" href="../reference/reference_phylo.html#ete3.PhyloTree" title="ete3.PhyloTree"><code class="xref py py-class docutils literal"><span class="pre">PhyloTree</span></code></a> instances allow molecular phylogenies to be linked
to the Multiple Sequence Alignments (MSA). To associate a MSA with a
phylogenetic tree you can use the <a class="reference internal" href="../reference/reference_phylo.html#ete3.PhyloNode.link_to_alignment" title="ete3.PhyloNode.link_to_alignment"><code class="xref py py-func docutils literal"><span class="pre">PhyloNode.link_to_alignment()</span></code></a>
method. You can use the <code class="xref py py-attr docutils literal"><span class="pre">alg_format</span></code> argument to specify its
format (See <a class="reference internal" href="../reference/reference_seqgroup.html#ete3.SeqGroup" title="ete3.SeqGroup"><code class="xref py py-class docutils literal"><span class="pre">SeqGroup</span></code></a> documentation for available formats)</p>
<p>Given that Fasta format are not only applicable for MSA but also for
<strong>Unaligned Sequences</strong>, you may also associate sequences of different
lengths with tree nodes.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">ete3</span> <span class="kn">import</span> <span class="n">PhyloTree</span>
<span class="n">fasta_txt</span> <span class="o">=</span> <span class="s">&quot;&quot;&quot;</span>
<span class="s">&gt;seqA</span>
<span class="s">MAEIPDETIQQFMALT---HNIAVQYLSEFGDLNEALNSYYASQTDDIKDRREEAH</span>
<span class="s">&gt;seqB</span>
<span class="s">MAEIPDATIQQFMALTNVSHNIAVQY--EFGDLNEALNSYYAYQTDDQKDRREEAH</span>
<span class="s">&gt;seqC</span>
<span class="s">MAEIPDATIQ---ALTNVSHNIAVQYLSEFGDLNEALNSYYASQTDDQPDRREEAH</span>
<span class="s">&gt;seqD</span>
<span class="s">MAEAPDETIQQFMALTNVSHNIAVQYLSEFGDLNEAL--------------REEAH</span>
<span class="s">&quot;&quot;&quot;</span>

<span class="c"># Load a tree and link it to an alignment.</span>
<span class="n">t</span> <span class="o">=</span> <span class="n">PhyloTree</span><span class="p">(</span><span class="s">&quot;(((seqA,seqB),seqC),seqD);&quot;</span><span class="p">)</span>
<span class="n">t</span><span class="o">.</span><span class="n">link_to_alignment</span><span class="p">(</span><span class="n">alignment</span><span class="o">=</span><span class="n">fasta_txt</span><span class="p">,</span> <span class="n">alg_format</span><span class="o">=</span><span class="s">&quot;fasta&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>The same could be done at the same time the tree is being loaded, by
using the <code class="xref py py-attr docutils literal"><span class="pre">alignment</span></code> and <code class="xref py py-attr docutils literal"><span class="pre">alg_format</span></code> arguments of
<a class="reference internal" href="../reference/reference_phylo.html#ete3.PhyloTree" title="ete3.PhyloTree"><code class="xref py py-class docutils literal"><span class="pre">PhyloTree</span></code></a>.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="c"># Load a tree and link it to an alignment.</span>
<span class="n">t</span> <span class="o">=</span> <span class="n">PhyloTree</span><span class="p">(</span><span class="s">&quot;(((seqA,seqB),seqC),seqD);&quot;</span><span class="p">,</span> <span class="n">alignment</span><span class="o">=</span><span class="n">fasta_txt</span><span class="p">,</span> <span class="n">alg_format</span><span class="o">=</span><span class="s">&quot;fasta&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>As currently implemented, sequence linking process is not strict,
which means that a perfect match between all node names and sequences
names <strong>is not required</strong>. Thus, if only one match is found between
sequences names within the MSA file and tree node names, only one tree
node will contain an associated sequence. Also, it is important to
note that sequence linking is not limited to terminal nodes. If
internal nodes are named, and such names find a match within the
provided MSA file, their corresponding sequences will be also loaded
into the tree structure. Once a MSA is linked, sequences will be
available for every tree node through its <code class="xref py py-attr docutils literal"><span class="pre">node.sequence</span></code>
attribute.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">ete3</span> <span class="kn">import</span> <span class="n">PhyloTree</span>
<span class="n">fasta_txt</span> <span class="o">=</span> <span class="s">&quot;&quot;&quot;</span>
<span class="s"> &gt;seqA</span>
<span class="s"> MAEIPDETIQQFMALT---HNIAVQYLSEFGDLNEALNSYYASQTDDIKDRREEAH</span>
<span class="s"> &gt;seqB</span>
<span class="s"> MAEIPDATIQQFMALTNVSHNIAVQY--EFGDLNEALNSYYAYQTDDQKDRREEAH</span>
<span class="s"> &gt;seqC</span>
<span class="s"> MAEIPDATIQ---ALTNVSHNIAVQYLSEFGDLNEALNSYYASQTDDQPDRREEAH</span>
<span class="s"> &gt;seqD</span>
<span class="s"> MAEAPDETIQQFMALTNVSHNIAVQYLSEFGDLNEAL--------------REEAH</span>
<span class="s">&quot;&quot;&quot;</span>
<span class="n">iphylip_txt</span> <span class="o">=</span> <span class="s">&quot;&quot;&quot;</span>
<span class="s"> 4 76</span>
<span class="s">      seqA   MAEIPDETIQ QFMALT---H NIAVQYLSEF GDLNEALNSY YASQTDDIKD RREEAHQFMA</span>
<span class="s">      seqB   MAEIPDATIQ QFMALTNVSH NIAVQY--EF GDLNEALNSY YAYQTDDQKD RREEAHQFMA</span>
<span class="s">      seqC   MAEIPDATIQ ---ALTNVSH NIAVQYLSEF GDLNEALNSY YASQTDDQPD RREEAHQFMA</span>
<span class="s">      seqD   MAEAPDETIQ QFMALTNVSH NIAVQYLSEF GDLNEAL--- ---------- -REEAHQ---</span>
<span class="s">             LTNVSHQFMA LTNVSH</span>
<span class="s">             LTNVSH---- ------</span>
<span class="s">             LTNVSH---- ------</span>
<span class="s">             -------FMA LTNVSH</span>
<span class="s">&quot;&quot;&quot;</span>
<span class="c"># Load a tree and link it to an alignment. As usual, &#39;alignment&#39; can</span>
<span class="c"># be the path to a file or data in text format.</span>
<span class="n">t</span> <span class="o">=</span> <span class="n">PhyloTree</span><span class="p">(</span><span class="s">&quot;(((seqA,seqB),seqC),seqD);&quot;</span><span class="p">,</span> <span class="n">alignment</span><span class="o">=</span><span class="n">fasta_txt</span><span class="p">,</span> <span class="n">alg_format</span><span class="o">=</span><span class="s">&quot;fasta&quot;</span><span class="p">)</span>

<span class="c">#We can now access the sequence of every leaf node</span>
<span class="k">print</span> <span class="s">&quot;These are the nodes and its sequences:&quot;</span>
<span class="k">for</span> <span class="n">leaf</span> <span class="ow">in</span> <span class="n">t</span><span class="o">.</span><span class="n">iter_leaves</span><span class="p">():</span>
    <span class="k">print</span> <span class="n">leaf</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="n">leaf</span><span class="o">.</span><span class="n">sequence</span>
<span class="c">#seqD MAEAPDETIQQFMALTNVSHNIAVQYLSEFGDLNEAL--------------REEAH</span>
<span class="c">#seqC MAEIPDATIQ---ALTNVSHNIAVQYLSEFGDLNEALNSYYASQTDDQPDRREEAH</span>
<span class="c">#seqA MAEIPDETIQQFMALT---HNIAVQYLSEFGDLNEALNSYYASQTDDIKDRREEAH</span>
<span class="c">#seqB MAEIPDATIQQFMALTNVSHNIAVQY--EFGDLNEALNSYYAYQTDDQKDRREEAH</span>
<span class="c">#</span>
<span class="c"># The associated alignment can be changed at any time</span>
<span class="n">t</span><span class="o">.</span><span class="n">link_to_alignment</span><span class="p">(</span><span class="n">alignment</span><span class="o">=</span><span class="n">iphylip_txt</span><span class="p">,</span> <span class="n">alg_format</span><span class="o">=</span><span class="s">&quot;iphylip&quot;</span><span class="p">)</span>
<span class="c"># Let&#39;s check that sequences have changed</span>
<span class="k">print</span> <span class="s">&quot;These are the nodes and its re-linked sequences:&quot;</span>
<span class="k">for</span> <span class="n">leaf</span> <span class="ow">in</span> <span class="n">t</span><span class="o">.</span><span class="n">iter_leaves</span><span class="p">():</span>
    <span class="k">print</span> <span class="n">leaf</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="n">leaf</span><span class="o">.</span><span class="n">sequence</span>
<span class="c">#seqD MAEAPDETIQQFMALTNVSHNIAVQYLSEFGDLNEAL--------------REEAHQ----------FMALTNVSH</span>
<span class="c">#seqC MAEIPDATIQ---ALTNVSHNIAVQYLSEFGDLNEALNSYYASQTDDQPDRREEAHQFMALTNVSH----------</span>
<span class="c">#seqA MAEIPDETIQQFMALT---HNIAVQYLSEFGDLNEALNSYYASQTDDIKDRREEAHQFMALTNVSHQFMALTNVSH</span>
<span class="c">#seqB MAEIPDATIQQFMALTNVSHNIAVQY--EFGDLNEALNSYYAYQTDDQKDRREEAHQFMALTNVSH----------</span>
<span class="c">#</span>
<span class="c"># The sequence attribute is considered as node feature, so you can</span>
<span class="c"># even include sequences in your extended newick format!</span>
<span class="k">print</span> <span class="n">t</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">features</span><span class="o">=</span><span class="p">[</span><span class="s">&quot;sequence&quot;</span><span class="p">],</span> <span class="n">format</span><span class="o">=</span><span class="mi">9</span><span class="p">)</span>
<span class="c">#</span>
<span class="c">#</span>
<span class="c"># (((seqA[&amp;&amp;NHX:sequence=MAEIPDETIQQFMALT---HNIAVQYLSEFGDLNEALNSYYASQTDDIKDRREEAHQF</span>
<span class="c"># MALTNVSHQFMALTNVSH],seqB[&amp;&amp;NHX:sequence=MAEIPDATIQQFMALTNVSHNIAVQY--EFGDLNEALNSY</span>
<span class="c"># YAYQTDDQKDRREEAHQFMALTNVSH----------]),seqC[&amp;&amp;NHX:sequence=MAEIPDATIQ---ALTNVSHNIA</span>
<span class="c"># VQYLSEFGDLNEALNSYYASQTDDQPDRREEAHQFMALTNVSH----------]),seqD[&amp;&amp;NHX:sequence=MAEAPD</span>
<span class="c"># ETIQQFMALTNVSHNIAVQYLSEFGDLNEAL--------------REEAHQ----------FMALTNVSH]);</span>
<span class="c">#</span>
<span class="c"># And yes, you can save this newick text and reload it into a PhyloTree instance.</span>
<span class="n">sametree</span> <span class="o">=</span> <span class="n">PhyloTree</span><span class="p">(</span><span class="n">t</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">features</span><span class="o">=</span><span class="p">[</span><span class="s">&quot;sequence&quot;</span><span class="p">]))</span>
<span class="k">print</span> <span class="s">&quot;Recovered tree with sequence features:&quot;</span>
<span class="k">print</span> <span class="n">sametree</span>
<span class="c">#</span>
<span class="c">#                              /-seqA</span>
<span class="c">#                    /--------|</span>
<span class="c">#          /--------|          \-seqB</span>
<span class="c">#         |         |</span>
<span class="c">#---------|          \-seqC</span>
<span class="c">#         |</span>
<span class="c">#          \-seqD</span>
<span class="c">#</span>
<span class="k">print</span> <span class="s">&quot;seqA sequence:&quot;</span><span class="p">,</span> <span class="p">(</span><span class="n">t</span><span class="o">&amp;</span><span class="s">&quot;seqA&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">sequence</span>
<span class="c"># MAEIPDETIQQFMALT---HNIAVQYLSEFGDLNEALNSYYASQTDDIKDRREEAHQFMALTNVSHQFMALTNVSH</span>
</pre></div>
</div>
</div>
<div class="section" id="visualization-of-phylogenetic-trees">
<h2><a class="toc-backref" href="#id4">Visualization of phylogenetic trees</a><a class="headerlink" href="#visualization-of-phylogenetic-trees" title="Permalink to this headline">Â¶</a></h2>
<p>PhyloTree instances can benefit from all the features of the
programmable drawing engine. However, a built-in phylogenetic layout
is provided for convenience.</p>
<p>All PhyloTree instances are, by default, attached to such layout for
tree visualization, thus allowing for in-place alignment visualization
and evolutionary events labeling.</p>
<div class="figure">
<img alt="../_images/phylotree.png" src="../_images/phylotree.png" />
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">ete3</span> <span class="kn">import</span> <span class="n">PhyloTree</span><span class="p">,</span> <span class="n">TreeStyle</span>

<span class="n">alg</span> <span class="o">=</span> <span class="s">&quot;&quot;&quot;</span>
<span class="s"> &gt;Dme_001</span>
<span class="s"> MAEIPDETIQQFMALT---HNIAVQYLSEFGDLNEAL--YYASQTDDIKDRREEAH</span>
<span class="s"> &gt;Dme_002</span>
<span class="s"> MAEIPDATIQQFMALTNVSHNIAVQY--EFGDLNEALNSYYAYQTDDQKDRREEAH</span>
<span class="s"> &gt;Cfa_001</span>
<span class="s"> MAEIPDATIQ---ALTNVSHNIAVQYLSEFGDLNEALNSYYASQTDDQPDRREEAH</span>
<span class="s"> &gt;Mms_001</span>
<span class="s"> MAEAPDETIQQFMALTNVSHNIAVQYLSEFGDLNEAL--------------REEAH</span>
<span class="s"> &gt;Hsa_001</span>
<span class="s"> MAEIPDETIQQFMALT---HNIAVQYLSEFGDLNEALNSYYASQTDDIKDRREEAH</span>
<span class="s"> &gt;Ptr_002</span>
<span class="s"> MAEIPDATIQ-FMALTNVSHNIAVQY--EFGDLNEALNSY--YQTDDQKDRREEAH</span>
<span class="s"> &gt;Mmu_002</span>
<span class="s"> MAEIPDATIQ---ALTNVSHNIAVQYLSEFGDLNEALNSYYASQTDDQPDRREEAH</span>
<span class="s"> &gt;Hsa_002</span>
<span class="s"> MAEAPDETIQQFM-LTNVSHNIAVQYLSEFGDLNEAL--------------REEAH</span>
<span class="s"> &gt;Mmu_001</span>
<span class="s"> MAEIPDETIQQFMALT---HNIAVQYLSEFGDLNEALNSYYASQTDDIKDRREEAH</span>
<span class="s"> &gt;Ptr_001</span>
<span class="s"> MAEIPDATIQ-FMALTNVSHNIAVQY--EFGDLNEALNSY--YQTDDQKDRREEAH</span>
<span class="s"> &gt;Mmu_001</span>
<span class="s"> MAEIPDATIQ---ALTNVSHNIAVQYLSEFGDLNEALNSYYASQTDDQPDRREEAH</span>
<span class="s">&quot;&quot;&quot;</span>

<span class="k">def</span> <span class="nf">get_example_tree</span><span class="p">():</span>

    <span class="c"># Performs a tree reconciliation analysis</span>
    <span class="n">gene_tree_nw</span> <span class="o">=</span> <span class="s">&#39;((Dme_001,Dme_002),(((Cfa_001,Mms_001),((Hsa_001,Ptr_001),Mmu_001)),(Ptr_002,(Hsa_002,Mmu_002))));&#39;</span>
    <span class="n">species_tree_nw</span> <span class="o">=</span> <span class="s">&quot;((((Hsa, Ptr), Mmu), (Mms, Cfa)), Dme);&quot;</span>
    <span class="n">genetree</span> <span class="o">=</span> <span class="n">PhyloTree</span><span class="p">(</span><span class="n">gene_tree_nw</span><span class="p">)</span>
    <span class="n">sptree</span> <span class="o">=</span> <span class="n">PhyloTree</span><span class="p">(</span><span class="n">species_tree_nw</span><span class="p">)</span>
    <span class="n">recon_tree</span><span class="p">,</span> <span class="n">events</span> <span class="o">=</span> <span class="n">genetree</span><span class="o">.</span><span class="n">reconcile</span><span class="p">(</span><span class="n">sptree</span><span class="p">)</span>
    <span class="n">recon_tree</span><span class="o">.</span><span class="n">link_to_alignment</span><span class="p">(</span><span class="n">alg</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">recon_tree</span><span class="p">,</span> <span class="n">TreeStyle</span><span class="p">()</span>

<span class="k">if</span> <span class="n">__name__</span> <span class="o">==</span> <span class="s">&quot;__main__&quot;</span><span class="p">:</span>
    <span class="c"># Visualize the reconciled tree</span>
    <span class="n">t</span><span class="p">,</span> <span class="n">ts</span> <span class="o">=</span> <span class="n">get_example_tree</span><span class="p">()</span>
    <span class="n">t</span><span class="o">.</span><span class="n">show</span><span class="p">(</span><span class="n">tree_style</span><span class="o">=</span><span class="n">ts</span><span class="p">)</span>
    <span class="c">#recon_tree.render(&quot;phylotree.png&quot;, w=750)</span>
</pre></div>
</div>
</div>
<div class="section" id="adding-taxonomic-information">
<h2><a class="toc-backref" href="#id5">Adding taxonomic information</a><a class="headerlink" href="#adding-taxonomic-information" title="Permalink to this headline">Â¶</a></h2>
<p id="taxonomic-info"><a class="reference internal" href="../reference/reference_phylo.html#ete3.PhyloTree" title="ete3.PhyloTree"><code class="xref py py-class docutils literal"><span class="pre">PhyloTree</span></code></a> instances allow to deal with leaf names and species
names separately.  This is useful when working with molecular
phylogenies, in which node names usually represent sequence
identifiers.  Species names will be stored in the
<a class="reference internal" href="../reference/reference_phylo.html#ete3.PhyloNode.species" title="ete3.PhyloNode.species"><code class="xref py py-attr docutils literal"><span class="pre">PhyloNode.species</span></code></a> attribute of each leaf node. The method
<a class="reference internal" href="../reference/reference_phylo.html#ete3.PhyloNode.get_species" title="ete3.PhyloNode.get_species"><code class="xref py py-func docutils literal"><span class="pre">PhyloNode.get_species()</span></code></a> can be used obtain the set of species
names found under a given internal node (speciation or duplication
event).  Often, sequence names do contain species information as a
part of the name, and ETE can parse this information automatically.</p>
<p>There are three ways to establish the species of the different tree
nodes:</p>
<ul class="simple">
<li>Default: The three first letters of node&#8217;s name represent the species</li>
<li>The species code of each node is dynamically created based on node&#8217;s name</li>
<li>The species code of each node is manually set.</li>
</ul>
<div class="section" id="automatic-control-of-species-info">
<h3><a class="toc-backref" href="#id6">Automatic control of species info</a><a class="headerlink" href="#automatic-control-of-species-info" title="Permalink to this headline">Â¶</a></h3>
<div class="highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">ete3</span> <span class="kn">import</span> <span class="n">PhyloTree</span>
<span class="c"># Reads a phylogenetic tree (using default species name encoding)</span>
<span class="n">t</span> <span class="o">=</span> <span class="n">PhyloTree</span><span class="p">(</span><span class="s">&quot;(((Hsa_001,Ptr_001),(Cfa_001,Mms_001)),(Dme_001,Dme_002));&quot;</span><span class="p">)</span>
<span class="c">#                              /-Hsa_001</span>
<span class="c">#                    /--------|</span>
<span class="c">#                   |          \-Ptr_001</span>
<span class="c">#          /--------|</span>
<span class="c">#         |         |          /-Cfa_001</span>
<span class="c">#         |          \--------|</span>
<span class="c">#---------|                    \-Mms_001</span>
<span class="c">#         |</span>
<span class="c">#         |          /-Dme_001</span>
<span class="c">#          \--------|</span>
<span class="c">#                    \-Dme_002</span>
<span class="c">#</span>
<span class="c"># Prints current leaf names and species codes</span>
<span class="k">print</span> <span class="s">&quot;Deafult mode:&quot;</span>
<span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="n">t</span><span class="o">.</span><span class="n">get_leaves</span><span class="p">():</span>
    <span class="k">print</span> <span class="s">&quot;node:&quot;</span><span class="p">,</span> <span class="n">n</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="s">&quot;Species name:&quot;</span><span class="p">,</span> <span class="n">n</span><span class="o">.</span><span class="n">species</span>
<span class="c"># node: Dme_001 Species name: Dme</span>
<span class="c"># node: Dme_002 Species name: Dme</span>
<span class="c"># node: Hsa_001 Species name: Hsa</span>
<span class="c"># node: Ptr_001 Species name: Ptr</span>
<span class="c"># node: Cfa_001 Species name: Cfa</span>
<span class="c"># node: Mms_001 Species name: Mms</span>
</pre></div>
</div>
</div>
<div class="section" id="automatic-and-custom-control-of-the-species-info">
<h3><a class="toc-backref" href="#id7">Automatic (and custom) control of the species info</a><a class="headerlink" href="#automatic-and-custom-control-of-the-species-info" title="Permalink to this headline">Â¶</a></h3>
<p>The default behavior can be changed by using the
<a class="reference internal" href="../reference/reference_phylo.html#ete3.PhyloNode.set_species_naming_function" title="ete3.PhyloNode.set_species_naming_function"><code class="xref py py-func docutils literal"><span class="pre">PhyloNode.set_species_naming_function()</span></code></a> method or by using the
<code class="xref py py-attr docutils literal"><span class="pre">sp_naming_function</span></code> argument of the <a class="reference internal" href="../reference/reference_phylo.html#ete3.PhyloTree" title="ete3.PhyloTree"><code class="xref py py-class docutils literal"><span class="pre">PhyloTree</span></code></a> class.
Note that, using the <code class="xref py py-attr docutils literal"><span class="pre">sp_naming_function</span></code> argument, the whole
tree structure will be initialized to use the provided parsing
function to obtain species name
information. <a class="reference internal" href="../reference/reference_phylo.html#ete3.PhyloNode.set_species_naming_function" title="ete3.PhyloNode.set_species_naming_function"><code class="xref py py-func docutils literal"><span class="pre">PhyloNode.set_species_naming_function()</span></code></a> (present in
all tree nodes) can be used to change the behavior in a previously
loaded tree, or to set different parsing function to different parts
of the tree.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">ete3</span> <span class="kn">import</span> <span class="n">PhyloTree</span>
<span class="c"># Reads a phylogenetic tree</span>
<span class="n">t</span> <span class="o">=</span> <span class="n">PhyloTree</span><span class="p">(</span><span class="s">&quot;(((Hsa_001,Ptr_001),(Cfa_001,Mms_001)),(Dme_001,Dme_002));&quot;</span><span class="p">)</span>

<span class="c"># Let&#39;s use our own leaf name parsing function to obtain species</span>
<span class="c"># names. All we need to do is create a python function that takes</span>
<span class="c"># node&#39;s name as argument and return its corresponding species name.</span>
<span class="k">def</span> <span class="nf">get_species_name</span><span class="p">(</span><span class="n">node_name_string</span><span class="p">):</span>
    <span class="c"># Species code is the first part of leaf name (separated by an</span>
    <span class="c">#  underscore character)</span>
    <span class="n">spcode</span> <span class="o">=</span> <span class="n">node_name_string</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&quot;_&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
    <span class="c"># We could even translate the code to complete names</span>
    <span class="n">code2name</span> <span class="o">=</span> <span class="p">{</span>
      <span class="s">&quot;Dme&quot;</span><span class="p">:</span><span class="s">&quot;Drosophila melanogaster&quot;</span><span class="p">,</span>
      <span class="s">&quot;Hsa&quot;</span><span class="p">:</span><span class="s">&quot;Homo sapiens&quot;</span><span class="p">,</span>
      <span class="s">&quot;Ptr&quot;</span><span class="p">:</span><span class="s">&quot;Pan troglodytes&quot;</span><span class="p">,</span>
      <span class="s">&quot;Mms&quot;</span><span class="p">:</span><span class="s">&quot;Mus musculus&quot;</span><span class="p">,</span>
      <span class="s">&quot;Cfa&quot;</span><span class="p">:</span><span class="s">&quot;Canis familiaris&quot;</span>
      <span class="p">}</span>
    <span class="k">return</span> <span class="n">code2name</span><span class="p">[</span><span class="n">spcode</span><span class="p">]</span>

<span class="c"># Now, let&#39;s ask the tree to use our custom species naming function</span>
<span class="n">t</span><span class="o">.</span><span class="n">set_species_naming_function</span><span class="p">(</span><span class="n">get_species_name</span><span class="p">)</span>
<span class="k">print</span> <span class="s">&quot;Custom mode:&quot;</span>
<span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="n">t</span><span class="o">.</span><span class="n">get_leaves</span><span class="p">():</span>
    <span class="k">print</span> <span class="s">&quot;node:&quot;</span><span class="p">,</span> <span class="n">n</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="s">&quot;Species name:&quot;</span><span class="p">,</span> <span class="n">n</span><span class="o">.</span><span class="n">species</span>

<span class="c"># node: Dme_001 Species name: Drosophila melanogaster</span>
<span class="c"># node: Dme_002 Species name: Drosophila melanogaster</span>
<span class="c"># node: Hsa_001 Species name: Homo sapiens</span>
<span class="c"># node: Ptr_001 Species name: Pan troglodytes</span>
<span class="c"># node: Cfa_001 Species name: Canis familiaris</span>
<span class="c"># node: Mms_001 Species name: Mus musculus</span>
</pre></div>
</div>
</div>
<div class="section" id="manual-control-of-the-species-info">
<h3><a class="toc-backref" href="#id8">Manual control of the species info</a><a class="headerlink" href="#manual-control-of-the-species-info" title="Permalink to this headline">Â¶</a></h3>
<p>To disable the automatic generation of species names based on node
names, a <code class="docutils literal"><span class="pre">None</span></code> value can be passed to the
<a class="reference internal" href="../reference/reference_phylo.html#ete3.PhyloNode.set_species_naming_function" title="ete3.PhyloNode.set_species_naming_function"><code class="xref py py-func docutils literal"><span class="pre">PhyloNode.set_species_naming_function()</span></code></a> function. From then on,
species attribute will not be automatically updated based on the name
of nodes and it could be controlled manually.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">ete3</span> <span class="kn">import</span> <span class="n">PhyloTree</span>
<span class="c"># Reads a phylogenetic tree</span>
<span class="n">t</span> <span class="o">=</span> <span class="n">PhyloTree</span><span class="p">(</span><span class="s">&quot;(((Hsa_001,Ptr_001),(Cfa_001,Mms_001)),(Dme_001,Dme_002));&quot;</span><span class="p">)</span>

<span class="c"># Of course, you can disable the automatic generation of species</span>
<span class="c"># names. To do so, you can set the species naming function to</span>
<span class="c"># None. This is useful to set the species names manually or for</span>
<span class="c"># reading them from a newick file. Other wise, species attribute would</span>
<span class="c"># be overwriten</span>
<span class="n">mynewick</span> <span class="o">=</span> <span class="s">&quot;&quot;&quot;</span>
<span class="s">(((Hsa_001[&amp;&amp;NHX:species=Human],Ptr_001[&amp;&amp;NHX:species=Chimp]),</span>
<span class="s">(Cfa_001[&amp;&amp;NHX:species=Dog],Mms_001[&amp;&amp;NHX:species=Mouse])),</span>
<span class="s">(Dme_001[&amp;&amp;NHX:species=Fly],Dme_002[&amp;&amp;NHX:species=Fly]));</span>
<span class="s">&quot;&quot;&quot;</span>
<span class="n">t</span> <span class="o">=</span> <span class="n">PhyloTree</span><span class="p">(</span><span class="n">mynewick</span><span class="p">,</span> <span class="n">sp_naming_function</span><span class="o">=</span><span class="bp">None</span><span class="p">)</span>
<span class="k">print</span> <span class="s">&quot;Disabled mode (manual set)&quot;</span>
<span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="n">t</span><span class="o">.</span><span class="n">get_leaves</span><span class="p">():</span>
    <span class="k">print</span> <span class="s">&quot;node:&quot;</span><span class="p">,</span> <span class="n">n</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="s">&quot;Species name:&quot;</span><span class="p">,</span> <span class="n">n</span><span class="o">.</span><span class="n">species</span>

<span class="c"># node: Dme_001 Species name: Fly</span>
<span class="c"># node: Dme_002 Species name: Fly</span>
<span class="c"># node: Hsa_001 Species name: Human</span>
<span class="c"># node: Ptr_001 Species name: Chimp</span>
<span class="c"># node: Cfa_001 Species name: Dog</span>
<span class="c"># node: Mms_001 Species name: Mouse</span>
</pre></div>
</div>
<p><strong>Full Example:</strong> <a class="reference download internal" href="../_downloads/species_aware_phylogenies.py"><code class="xref download docutils literal"><span class="pre">Species</span> <span class="pre">aware</span> <span class="pre">trees</span></code></a>.</p>
</div>
</div>
<div class="section" id="detecting-evolutionary-events">
<h2><a class="toc-backref" href="#id9">Detecting evolutionary events</a><a class="headerlink" href="#detecting-evolutionary-events" title="Permalink to this headline">Â¶</a></h2>
<p>There are several ways to automatically detect duplication and
speciation nodes. ETE provides two methodologies: One implements the
algorithm described in <a class="reference external" href="http://genomebiology.com/2007/8/6/R109">Huerta-Cepas (2007)</a> and is based on the species
overlap (SO) between partitions and thus does not depend on the
availability of a species tree. The second, which requires the
comparison between the gene tree and a previously defined species
tree, implements a strict tree reconciliation algorithm (Page and
Charleston, 1997). By detecting evolutionary events, orthology and
paralogy relationships among sequences can also be inferred.  Find a
comparison of both methods in <a class="reference external" href="http://www.plosone.org/article/info:doi%2F10.1371%2Fjournal.pone.0004357">Marcet-Houben and Gabaldon (2009)</a>.</p>
<div class="section" id="species-overlap-so-algorithm">
<span id="spoverlap"></span><h3><a class="toc-backref" href="#id10">Species Overlap (SO) algorithm</a><a class="headerlink" href="#species-overlap-so-algorithm" title="Permalink to this headline">Â¶</a></h3>
<p>In order to apply the SO algorithm, you can use the
<a class="reference internal" href="../reference/reference_phylo.html#ete3.PhyloNode.get_descendant_evol_events" title="ete3.PhyloNode.get_descendant_evol_events"><code class="xref py py-func docutils literal"><span class="pre">PhyloNode.get_descendant_evol_events()</span></code></a> method (it will detect
all evolutionary events under the current node) or the
<a class="reference internal" href="../reference/reference_phylo.html#ete3.PhyloNode.get_my_evol_events" title="ete3.PhyloNode.get_my_evol_events"><code class="xref py py-func docutils literal"><span class="pre">PhyloNode.get_my_evol_events()</span></code></a> method (it will detect only the
evolutionary events in which current node, a leaf, is involved).</p>
<p>By default the <strong>species overlap score (SOS) threshold</strong> is set to
0.0, which means that a single species in common between two node
branches will rise a duplication event. This has been shown to perform
the best with real data, however you can adjust the threshold using
the <code class="docutils literal"><span class="pre">sos_thr</span></code> argument present in both methods.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">ete3</span> <span class="kn">import</span> <span class="n">PhyloTree</span>
<span class="c"># Loads an example tree</span>
<span class="n">nw</span> <span class="o">=</span> <span class="s">&quot;&quot;&quot;</span>
<span class="s">((Dme_001,Dme_002),(((Cfa_001,Mms_001),((Hsa_001,Ptr_001),Mmu_001)),</span>
<span class="s">(Ptr_002,(Hsa_002,Mmu_002))));</span>
<span class="s">&quot;&quot;&quot;</span>
<span class="n">t</span> <span class="o">=</span> <span class="n">PhyloTree</span><span class="p">(</span><span class="n">nw</span><span class="p">)</span>
<span class="k">print</span> <span class="n">t</span>
<span class="c">#                    /-Dme_001</span>
<span class="c">#          /--------|</span>
<span class="c">#         |          \-Dme_002</span>
<span class="c">#         |</span>
<span class="c">#         |                              /-Cfa_001</span>
<span class="c">#         |                    /--------|</span>
<span class="c">#---------|                   |          \-Mms_001</span>
<span class="c">#         |          /--------|</span>
<span class="c">#         |         |         |                    /-Hsa_001</span>
<span class="c">#         |         |         |          /--------|</span>
<span class="c">#         |         |          \--------|          \-Ptr_001</span>
<span class="c">#          \--------|                   |</span>
<span class="c">#                   |                    \-Mmu_001</span>
<span class="c">#                   |</span>
<span class="c">#                   |          /-Ptr_002</span>
<span class="c">#                    \--------|</span>
<span class="c">#                             |          /-Hsa_002</span>
<span class="c">#                              \--------|</span>
<span class="c">#                                        \-Mmu_002</span>
<span class="c">#</span>
<span class="c"># To obtain all the evolutionary events involving a given leaf node we</span>
<span class="c"># use get_my_evol_events method</span>
<span class="n">matches</span> <span class="o">=</span> <span class="n">t</span><span class="o">.</span><span class="n">search_nodes</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s">&quot;Hsa_001&quot;</span><span class="p">)</span>
<span class="n">human_seq</span> <span class="o">=</span> <span class="n">matches</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="c"># Obtains its evolutionary events</span>
<span class="n">events</span> <span class="o">=</span> <span class="n">human_seq</span><span class="o">.</span><span class="n">get_my_evol_events</span><span class="p">()</span>
<span class="c"># Print its orthology and paralogy relationships</span>
<span class="k">print</span> <span class="s">&quot;Events detected that involve Hsa_001:&quot;</span>
<span class="k">for</span> <span class="n">ev</span> <span class="ow">in</span> <span class="n">events</span><span class="p">:</span>
    <span class="k">if</span> <span class="n">ev</span><span class="o">.</span><span class="n">etype</span> <span class="o">==</span> <span class="s">&quot;S&quot;</span><span class="p">:</span>
        <span class="k">print</span> <span class="s">&#39;   ORTHOLOGY RELATIONSHIP:&#39;</span><span class="p">,</span> <span class="s">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">ev</span><span class="o">.</span><span class="n">in_seqs</span><span class="p">),</span> <span class="s">&quot;&lt;====&gt;&quot;</span><span class="p">,</span> <span class="s">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">ev</span><span class="o">.</span><span class="n">out_seqs</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">ev</span><span class="o">.</span><span class="n">etype</span> <span class="o">==</span> <span class="s">&quot;D&quot;</span><span class="p">:</span>
        <span class="k">print</span> <span class="s">&#39;   PARALOGY RELATIONSHIP:&#39;</span><span class="p">,</span> <span class="s">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">ev</span><span class="o">.</span><span class="n">in_seqs</span><span class="p">),</span> <span class="s">&quot;&lt;====&gt;&quot;</span><span class="p">,</span> <span class="s">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">ev</span><span class="o">.</span><span class="n">out_seqs</span><span class="p">)</span>

<span class="c"># Alternatively, you can scan the whole tree topology</span>
<span class="n">events</span> <span class="o">=</span> <span class="n">t</span><span class="o">.</span><span class="n">get_descendant_evol_events</span><span class="p">()</span>
<span class="c"># Print its orthology and paralogy relationships</span>
<span class="k">print</span> <span class="s">&quot;Events detected from the root of the tree&quot;</span>
<span class="k">for</span> <span class="n">ev</span> <span class="ow">in</span> <span class="n">events</span><span class="p">:</span>
    <span class="k">if</span> <span class="n">ev</span><span class="o">.</span><span class="n">etype</span> <span class="o">==</span> <span class="s">&quot;S&quot;</span><span class="p">:</span>
        <span class="k">print</span> <span class="s">&#39;   ORTHOLOGY RELATIONSHIP:&#39;</span><span class="p">,</span> <span class="s">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">ev</span><span class="o">.</span><span class="n">in_seqs</span><span class="p">),</span> <span class="s">&quot;&lt;====&gt;&quot;</span><span class="p">,</span> <span class="s">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">ev</span><span class="o">.</span><span class="n">out_seqs</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">ev</span><span class="o">.</span><span class="n">etype</span> <span class="o">==</span> <span class="s">&quot;D&quot;</span><span class="p">:</span>
        <span class="k">print</span> <span class="s">&#39;   PARALOGY RELATIONSHIP:&#39;</span><span class="p">,</span> <span class="s">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">ev</span><span class="o">.</span><span class="n">in_seqs</span><span class="p">),</span> <span class="s">&quot;&lt;====&gt;&quot;</span><span class="p">,</span> <span class="s">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">ev</span><span class="o">.</span><span class="n">out_seqs</span><span class="p">)</span>

<span class="c"># If we are only interested in the orthology and paralogy relationship</span>
<span class="c"># among a given set of species, we can filter the list of sequences</span>
<span class="c">#</span>
<span class="c"># fseqs is a function that, given a list of sequences, returns only</span>
<span class="c"># those from human and mouse</span>
<span class="n">fseqs</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">slist</span><span class="p">:</span> <span class="p">[</span><span class="n">s</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">slist</span> <span class="k">if</span> <span class="n">s</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s">&quot;Hsa&quot;</span><span class="p">)</span> <span class="ow">or</span> <span class="n">s</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s">&quot;Mms&quot;</span><span class="p">)]</span>
<span class="k">print</span> <span class="s">&quot;Paralogy relationships among human and mouse&quot;</span>
<span class="k">for</span> <span class="n">ev</span> <span class="ow">in</span> <span class="n">events</span><span class="p">:</span>
    <span class="k">if</span> <span class="n">ev</span><span class="o">.</span><span class="n">etype</span> <span class="o">==</span> <span class="s">&quot;D&quot;</span><span class="p">:</span>
        <span class="c"># Prints paralogy relationships considering only human and</span>
        <span class="c"># mouse. Some duplication event may not involve such species,</span>
        <span class="c"># so they will be empty</span>
        <span class="k">print</span> <span class="s">&#39;   PARALOGY RELATIONSHIP:&#39;</span><span class="p">,</span> \
            <span class="s">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">fseqs</span><span class="p">(</span><span class="n">ev</span><span class="o">.</span><span class="n">in_seqs</span><span class="p">)),</span> \
            <span class="s">&quot;&lt;====&gt;&quot;</span><span class="p">,</span>\
            <span class="s">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">fseqs</span><span class="p">(</span><span class="n">ev</span><span class="o">.</span><span class="n">out_seqs</span><span class="p">))</span>

<span class="c"># Note that besides the list of events returned, the detection</span>
<span class="c"># algorithm has labeled the tree nodes according with the</span>
<span class="c"># predictions. We can use such lables as normal node features.</span>
<span class="n">dups</span> <span class="o">=</span> <span class="n">t</span><span class="o">.</span><span class="n">search_nodes</span><span class="p">(</span><span class="n">evoltype</span><span class="o">=</span><span class="s">&quot;D&quot;</span><span class="p">)</span> <span class="c"># Return all duplication nodes</span>
</pre></div>
</div>
</div>
<div class="section" id="tree-reconciliation-algorithm">
<h3><a class="toc-backref" href="#id11">Tree reconciliation algorithm</a><a class="headerlink" href="#tree-reconciliation-algorithm" title="Permalink to this headline">Â¶</a></h3>
<p>Tree reconciliation algorithm uses a predefined species tree to infer
all the necessary genes losses that explain a given gene tree
topology. Consequently, duplication and separation nodes will strictly
follow the species tree topology.</p>
<p>To perform a tree reconciliation analysis over a given node in a
molecular phylogeny you can use the <a class="reference internal" href="../reference/reference_phylo.html#ete3.PhyloNode.reconcile" title="ete3.PhyloNode.reconcile"><code class="xref py py-func docutils literal"><span class="pre">PhyloNode.reconcile()</span></code></a>
method, which requires a species <a class="reference internal" href="../reference/reference_phylo.html#ete3.PhyloTree" title="ete3.PhyloTree"><code class="xref py py-class docutils literal"><span class="pre">PhyloTree</span></code></a> as its first
argument. Leaf node names in the the species are expected to be the
same species codes in the gene tree (see
<a class="reference internal" href="#taxonomic-info">taxonomic_info</a>). All species codes present in the
gene tree should appear in the species tree.</p>
<p>As a result, the <a class="reference internal" href="../reference/reference_phylo.html#ete3.PhyloNode.reconcile" title="ete3.PhyloNode.reconcile"><code class="xref py py-func docutils literal"><span class="pre">PhyloNode.reconcile()</span></code></a> method will label the
original gene tree nodes as duplication or speciation, will return the
list of inferred events, and will return a new <strong>reconcilied tree</strong>
(<a class="reference internal" href="../reference/reference_phylo.html#ete3.PhyloTree" title="ete3.PhyloTree"><code class="xref py py-class docutils literal"><span class="pre">PhyloTree</span></code></a> instance), in which inferred gene losses are
present and labeled.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">ete3</span> <span class="kn">import</span> <span class="n">PhyloTree</span>

<span class="c"># Loads a gene tree and its corresponding species tree. Note that</span>
<span class="c"># species names in sptree are the 3 firs letters of leaf nodes in</span>
<span class="c"># genetree.</span>
<span class="n">gene_tree_nw</span> <span class="o">=</span> <span class="s">&#39;((Dme_001,Dme_002),(((Cfa_001,Mms_001),((Hsa_001,Ptr_001),Mmu_001)),(Ptr_002,(Hsa_002,Mmu_002))));&#39;</span>
<span class="n">species_tree_nw</span> <span class="o">=</span> <span class="s">&quot;((((Hsa, Ptr), Mmu), (Mms, Cfa)), Dme);&quot;</span>
<span class="n">genetree</span> <span class="o">=</span> <span class="n">PhyloTree</span><span class="p">(</span><span class="n">gene_tree_nw</span><span class="p">)</span>
<span class="n">sptree</span> <span class="o">=</span> <span class="n">PhyloTree</span><span class="p">(</span><span class="n">species_tree_nw</span><span class="p">)</span>
<span class="k">print</span> <span class="n">genetree</span>
<span class="c">#                    /-Dme_001</span>
<span class="c">#          /--------|</span>
<span class="c">#         |          \-Dme_002</span>
<span class="c">#         |</span>
<span class="c">#         |                              /-Cfa_001</span>
<span class="c">#         |                    /--------|</span>
<span class="c">#---------|                   |          \-Mms_001</span>
<span class="c">#         |          /--------|</span>
<span class="c">#         |         |         |                    /-Hsa_001</span>
<span class="c">#         |         |         |          /--------|</span>
<span class="c">#         |         |          \--------|          \-Ptr_001</span>
<span class="c">#          \--------|                   |</span>
<span class="c">#                   |                    \-Mmu_001</span>
<span class="c">#                   |</span>
<span class="c">#                   |          /-Ptr_002</span>
<span class="c">#                    \--------|</span>
<span class="c">#                             |          /-Hsa_002</span>
<span class="c">#                              \--------|</span>
<span class="c">#                                        \-Mmu_002</span>
<span class="c">#</span>
<span class="c"># Let&#39;s reconcile our genetree with the species tree</span>
<span class="n">recon_tree</span><span class="p">,</span> <span class="n">events</span> <span class="o">=</span> <span class="n">genetree</span><span class="o">.</span><span class="n">reconcile</span><span class="p">(</span><span class="n">sptree</span><span class="p">)</span>
<span class="c"># a new &quot;reconcilied tree&quot; is returned. As well as the list of</span>
<span class="c"># inferred events.</span>
<span class="k">print</span> <span class="s">&quot;Orthology and Paralogy relationships:&quot;</span>
<span class="k">for</span> <span class="n">ev</span> <span class="ow">in</span> <span class="n">events</span><span class="p">:</span>
    <span class="k">if</span> <span class="n">ev</span><span class="o">.</span><span class="n">etype</span> <span class="o">==</span> <span class="s">&quot;S&quot;</span><span class="p">:</span>
        <span class="k">print</span> <span class="s">&#39;ORTHOLOGY RELATIONSHIP:&#39;</span><span class="p">,</span> <span class="s">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">ev</span><span class="o">.</span><span class="n">inparalogs</span><span class="p">),</span> <span class="s">&quot;&lt;====&gt;&quot;</span><span class="p">,</span> <span class="s">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">ev</span><span class="o">.</span><span class="n">orthologs</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">ev</span><span class="o">.</span><span class="n">etype</span> <span class="o">==</span> <span class="s">&quot;D&quot;</span><span class="p">:</span>
        <span class="k">print</span> <span class="s">&#39;PARALOGY RELATIONSHIP:&#39;</span><span class="p">,</span> <span class="s">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">ev</span><span class="o">.</span><span class="n">inparalogs</span><span class="p">),</span> <span class="s">&quot;&lt;====&gt;&quot;</span><span class="p">,</span> <span class="s">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">ev</span><span class="o">.</span><span class="n">outparalogs</span><span class="p">)</span>
<span class="c"># And we can explore the resulting reconciled tree</span>
<span class="k">print</span> <span class="n">recon_tree</span>
<span class="c"># You will notice how the reconcilied tree is the same as the gene</span>
<span class="c"># tree with some added branches. They are inferred gene losses.</span>
<span class="c">#</span>
<span class="c">#</span>
<span class="c">#                    /-Dme_001</span>
<span class="c">#          /--------|</span>
<span class="c">#         |          \-Dme_002</span>
<span class="c">#         |</span>
<span class="c">#         |                              /-Cfa_001</span>
<span class="c">#         |                    /--------|</span>
<span class="c">#         |                   |          \-Mms_001</span>
<span class="c">#---------|          /--------|</span>
<span class="c">#         |         |         |                    /-Hsa_001</span>
<span class="c">#         |         |         |          /--------|</span>
<span class="c">#         |         |          \--------|          \-Ptr_001</span>
<span class="c">#         |         |                   |</span>
<span class="c">#         |         |                    \-Mmu_001</span>
<span class="c">#          \--------|</span>
<span class="c">#                   |                    /-Mms</span>
<span class="c">#                   |          /--------|</span>
<span class="c">#                   |         |          \-Cfa</span>
<span class="c">#                   |         |</span>
<span class="c">#                   |         |                              /-Hsa</span>
<span class="c">#                    \--------|                    /--------|</span>
<span class="c">#                             |          /--------|          \-Ptr_002</span>
<span class="c">#                             |         |         |</span>
<span class="c">#                             |         |          \-Mmu</span>
<span class="c">#                              \--------|</span>
<span class="c">#                                       |                    /-Ptr</span>
<span class="c">#                                       |          /--------|</span>
<span class="c">#                                        \--------|          \-Hsa_002</span>
<span class="c">#                                                 |</span>
<span class="c">#                                                  \-Mmu_002</span>
<span class="c">#</span>
<span class="c"># And we can visualize the trees using the default phylogeny</span>
<span class="c"># visualization layout</span>
<span class="n">genetree</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
<span class="n">recon_tree</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="section" id="a-closer-look-to-the-evolutionary-event-object">
<h3><a class="toc-backref" href="#id12">A closer look to the evolutionary event object</a><a class="headerlink" href="#a-closer-look-to-the-evolutionary-event-object" title="Permalink to this headline">Â¶</a></h3>
<p>Both methods, species overlap and tree reconciliation, can be used to
label each tree node as a duplication or speciation event. Thus, the
<code class="xref py py-attr docutils literal"><span class="pre">PhyloNode.evoltype</span></code> attribute of every node will be set to one
of the following states: <code class="docutils literal"><span class="pre">D</span></code> (Duplication), <code class="docutils literal"><span class="pre">S</span></code> (Speciation) or
<code class="docutils literal"><span class="pre">L</span></code> gene loss.</p>
<p>Additionally, a list of all the detected events is returned. Each
event is a python object of type <a class="reference internal" href="../reference/reference_phylo.html#ete3.phylo.EvolEvent" title="ete3.phylo.EvolEvent"><code class="xref py py-class docutils literal"><span class="pre">phylo.EvolEvent</span></code></a>, containing
some basic information about each event ( <code class="xref py py-attr docutils literal"><span class="pre">etype</span></code>,
<code class="xref py py-attr docutils literal"><span class="pre">in_seqs</span></code>, <code class="xref py py-attr docutils literal"><span class="pre">out_seqs</span></code>, <code class="xref py py-attr docutils literal"><span class="pre">node</span></code>):</p>
<p>If an event represents a duplication, <code class="docutils literal"><span class="pre">in_seqs</span></code> <strong>are all
paralogous</strong> to <code class="docutils literal"><span class="pre">out_seqs</span></code>. Similarly, if an event represents a
speciation, <code class="docutils literal"><span class="pre">in_seqs</span></code> <strong>are all orthologous</strong> to <code class="docutils literal"><span class="pre">out_seqs</span></code>.</p>
</div>
</div>
<div class="section" id="relative-dating-phylogenetic-nodes">
<h2><a class="toc-backref" href="#id13">Relative dating phylogenetic nodes</a><a class="headerlink" href="#relative-dating-phylogenetic-nodes" title="Permalink to this headline">Â¶</a></h2>
<p>In molecular phylogeny, nodes can be interpreted as evolutionary
events. Therefore, they represent duplication or speciation events. In
the case of gene duplication events, nodes can also be assigned to a
certain point in a relative temporal scale. In other words, you can
obtain a relative dating of all the duplication events detected.</p>
<p>Although <strong>absolute dating is always preferred and more precise</strong>,
topological dating provides a faster approach to compare the relative
age of paralogous sequences (<a class="reference external" href="http://bioinformatics.oxfordjournals.org/content/27/1/38.long">read this</a> for
a comparison with other methods, such as the use of synonymous
substitution rates as a proxy to the divergence time).</p>
<p>Some applications of topological dating can be found in <a class="reference external" href="http://genomebiology.com/2007/8/6/r109">Huerta-Cepas
et al, 2007</a> or, more
recently, in <a class="reference external" href="http://bib.oxfordjournals.org/content/12/5/442.abstract">Huerta-Cepas et al, 2011</a> or
<a class="reference external" href="http://www.nature.com/nature/journal/v468/n7325/full/nature09634.html">Kalinka et al, 2001</a>.</p>
<div class="section" id="implementation">
<h3><a class="toc-backref" href="#id14">Implementation</a><a class="headerlink" href="#implementation" title="Permalink to this headline">Â¶</a></h3>
<p>The aim of relative dating is to establish a gradient of ages among
sequences. For this, a reference species needs to be fixed, so the
gradient of ages will be referred to that referent point.</p>
<p>Thus, if our reference species is <cite>Human</cite>, we could establish the
following gradient of species:</p>
<ul class="simple">
<li>(1) Human -&gt; (2) Other Primates -&gt; (3) Mammals -&gt; (4) Vertebrates</li>
</ul>
<p>So, nodes in a tree can be assigned to one of the above categories
depending on the sequences grouped. For instance:</p>
<ul class="simple">
<li>A node with only human sequences will be mapped to (1).</li>
<li>A node with human and orangutan sequences will be mapped to (2)</li>
<li>A node with human a fish sequences will be mapped to (4)</li>
</ul>
<p>This simple calculation can be done automatically by encoding the
gradient of species ages as Python dictionary.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">relative_dist</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s">&quot;human&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span> <span class="c"># human</span>
    <span class="s">&quot;chimp&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span> <span class="c"># Primates non human</span>
    <span class="s">&quot;rat&quot;</span><span class="p">:</span>   <span class="mi">2</span><span class="p">,</span> <span class="c"># Mammals non primates</span>
    <span class="s">&quot;mouse&quot;</span><span class="p">:</span> <span class="mi">2</span><span class="p">,</span> <span class="c"># Mammals non primates</span>
    <span class="s">&quot;fish&quot;</span><span class="p">:</span>  <span class="mi">3</span>  <span class="c"># Vertebrates non mammals</span>
    <span class="p">}</span>
</pre></div>
</div>
<p>Once done, ETE can check the relative age of any tree node.  The
<a class="reference internal" href="../reference/reference_phylo.html#ete3.PhyloNode.get_age" title="ete3.PhyloNode.get_age"><code class="xref py py-func docutils literal"><span class="pre">PhyloNode.get_age()</span></code></a> method can be used to that purpose.</p>
<p>For example, let&#8217;s consider the following gene tree:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="c">#                         /-humanA</span>
<span class="c">#                    /---|</span>
<span class="c">#                   |     \-chimpA</span>
<span class="c">#               /Dup1</span>
<span class="c">#              |    |     /-humanB</span>
<span class="c">#          /---|     \---|</span>
<span class="c">#         |    |          \-chimpB</span>
<span class="c">#     /---|    |</span>
<span class="c">#    |    |     \-mouseA</span>
<span class="c">#    |    |</span>
<span class="c">#    |     \-fish</span>
<span class="c">#-Dup3</span>
<span class="c">#    |               /-humanC</span>
<span class="c">#    |          /---|</span>
<span class="c">#    |     /---|     \-chimpC</span>
<span class="c">#    |    |    |</span>
<span class="c">#     \Dup2     \-humanD</span>
<span class="c">#         |</span>
<span class="c">#         |     /-ratC</span>
<span class="c">#          \---|</span>
<span class="c">#               \-mouseC</span>
</pre></div>
</div>
<p>the expected node dating would be:</p>
<blockquote>
<div><ul class="simple">
<li>Dup1 will be assigned to primates (most distant species is
chimp). <code class="docutils literal"><span class="pre">Dup1.get_age(relative_distances)</span></code> will return 1</li>
<li>Dup2 will be assigned to mammals [2] (most distant species are rat
and mouse). <code class="docutils literal"><span class="pre">Dup2.get_age(relative_distances)</span></code> will return 2</li>
<li>Dup3 will be assigned to mammals [3] (most distant species is
fish). <code class="docutils literal"><span class="pre">Dup3.get_age(relative_distances)</span></code> will return 3</li>
</ul>
</div></blockquote>
<div class="highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">ete3</span> <span class="kn">import</span> <span class="n">PhyloTree</span>
<span class="c"># Creates a gene phylogeny with several duplication events at</span>
<span class="c"># different levels. Note that we are using the default method for</span>
<span class="c"># detecting the species code of leaves (three first lettes in the node</span>
<span class="c"># name are considered the species code).</span>
<span class="n">nw</span> <span class="o">=</span> <span class="s">&quot;&quot;&quot;</span>
<span class="s">((Dme_001,Dme_002),(((Cfa_001,Mms_001),((((Hsa_001,Hsa_003),Ptr_001)</span>
<span class="s">,Mmu_001),((Hsa_004,Ptr_004),Mmu_004))),(Ptr_002,(Hsa_002,Mmu_002))));</span>
<span class="s">&quot;&quot;&quot;</span>
<span class="n">t</span> <span class="o">=</span> <span class="n">PhyloTree</span><span class="p">(</span><span class="n">nw</span><span class="p">)</span>
<span class="k">print</span> <span class="s">&quot;Original tree:&quot;</span><span class="p">,</span>
<span class="k">print</span> <span class="n">t</span>
<span class="c">#</span>
<span class="c">#             /-Dme_001</span>
<span class="c">#   /--------|</span>
<span class="c">#  |          \-Dme_002</span>
<span class="c">#  |</span>
<span class="c">#  |                              /-Cfa_001</span>
<span class="c">#  |                    /--------|</span>
<span class="c">#  |                   |          \-Mms_001</span>
<span class="c">#  |                   |</span>
<span class="c">#--|                   |                                        /-Hsa_001</span>
<span class="c">#  |                   |                              /--------|</span>
<span class="c">#  |          /--------|                    /--------|          \-Hsa_003</span>
<span class="c">#  |         |         |                   |         |</span>
<span class="c">#  |         |         |          /--------|          \-Ptr_001</span>
<span class="c">#  |         |         |         |         |</span>
<span class="c">#  |         |         |         |          \-Mmu_001</span>
<span class="c">#  |         |          \--------|</span>
<span class="c">#   \--------|                   |                    /-Hsa_004</span>
<span class="c">#            |                   |          /--------|</span>
<span class="c">#            |                    \--------|          \-Ptr_004</span>
<span class="c">#            |                             |</span>
<span class="c">#            |                              \-Mmu_004</span>
<span class="c">#            |</span>
<span class="c">#            |          /-Ptr_002</span>
<span class="c">#             \--------|</span>
<span class="c">#                      |          /-Hsa_002</span>
<span class="c">#                       \--------|</span>
<span class="c">#                                 \-Mmu_002</span>
<span class="c"># Create a dictionary with relative ages for the species present in</span>
<span class="c"># the phylogenetic tree.  Note that ages are only relative numbers to</span>
<span class="c"># define which species are older, and that different species can</span>
<span class="c"># belong to the same age.</span>
<span class="n">species2age</span> <span class="o">=</span> <span class="p">{</span>
  <span class="s">&#39;Hsa&#39;</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span> <span class="c"># Homo sapiens (Hominids)</span>
  <span class="s">&#39;Ptr&#39;</span><span class="p">:</span> <span class="mi">2</span><span class="p">,</span> <span class="c"># P. troglodytes (primates)</span>
  <span class="s">&#39;Mmu&#39;</span><span class="p">:</span> <span class="mi">2</span><span class="p">,</span> <span class="c"># Macaca mulata (primates)</span>
  <span class="s">&#39;Mms&#39;</span><span class="p">:</span> <span class="mi">3</span><span class="p">,</span> <span class="c"># Mus musculus (mammals)</span>
  <span class="s">&#39;Cfa&#39;</span><span class="p">:</span> <span class="mi">3</span><span class="p">,</span> <span class="c"># Canis familiaris (mammals)</span>
  <span class="s">&#39;Dme&#39;</span><span class="p">:</span> <span class="mi">4</span>  <span class="c"># Drosophila melanogaster (metazoa)</span>
<span class="p">}</span>
<span class="c"># We can translate each number to its correspondig taxonomic number</span>
<span class="n">age2name</span> <span class="o">=</span> <span class="p">{</span>
  <span class="mi">1</span><span class="p">:</span><span class="s">&quot;hominids&quot;</span><span class="p">,</span>
  <span class="mi">2</span><span class="p">:</span><span class="s">&quot;primates&quot;</span><span class="p">,</span>
  <span class="mi">3</span><span class="p">:</span><span class="s">&quot;mammals&quot;</span><span class="p">,</span>
  <span class="mi">4</span><span class="p">:</span><span class="s">&quot;metazoa&quot;</span>
<span class="p">}</span>
<span class="n">event1</span><span class="o">=</span> <span class="n">t</span><span class="o">.</span><span class="n">get_common_ancestor</span><span class="p">(</span><span class="s">&quot;Hsa_001&quot;</span><span class="p">,</span> <span class="s">&quot;Hsa_004&quot;</span><span class="p">)</span>
<span class="n">event2</span><span class="o">=</span><span class="n">t</span><span class="o">.</span><span class="n">get_common_ancestor</span><span class="p">(</span><span class="s">&quot;Hsa_001&quot;</span><span class="p">,</span> <span class="s">&quot;Hsa_002&quot;</span><span class="p">)</span>
<span class="k">print</span>
<span class="k">print</span> <span class="s">&quot;The duplication event leading to the human sequences Hsa_001 and &quot;</span><span class="o">+</span>\
    <span class="s">&quot;Hsa_004 is dated at: &quot;</span><span class="p">,</span> <span class="n">age2name</span><span class="p">[</span><span class="n">event1</span><span class="o">.</span><span class="n">get_age</span><span class="p">(</span><span class="n">species2age</span><span class="p">)]</span>
<span class="k">print</span> <span class="s">&quot;The duplication event leading to the human sequences Hsa_001 and &quot;</span><span class="o">+</span>\
    <span class="s">&quot;Hsa_002 is dated at: &quot;</span><span class="p">,</span> <span class="n">age2name</span><span class="p">[</span><span class="n">event2</span><span class="o">.</span><span class="n">get_age</span><span class="p">(</span><span class="n">species2age</span><span class="p">)]</span>
<span class="c"># The duplication event leading to the human sequences Hsa_001 and Hsa_004</span>
<span class="c"># is dated at:  primates</span>
<span class="c">#</span>
<span class="c"># The duplication event leading to the human sequences Hsa_001 and Hsa_002</span>
<span class="c"># is dated at:  mammals</span>
</pre></div>
</div>
<div class="admonition warning">
<p class="first admonition-title">Warning</p>
<p class="last">Note that relative distances will vary depending on your reference
species.</p>
</div>
</div>
</div>
<div class="section" id="automatic-rooting-outgroup-detection">
<h2><a class="toc-backref" href="#id15">Automatic rooting (outgroup detection)</a><a class="headerlink" href="#automatic-rooting-outgroup-detection" title="Permalink to this headline">Â¶</a></h2>
<p>Two methods are provided to assist in the automatic rooting of
phylogenetic trees. Since tree nodes contain relative age information
(based on the species code autodetection), the same relative age
dictionaries can be used to detect the farthest and oldest node in a
tree to given sequences.</p>
<p><a class="reference internal" href="../reference/reference_phylo.html#ete3.PhyloNode.get_farthest_oldest_node" title="ete3.PhyloNode.get_farthest_oldest_node"><code class="xref py py-func docutils literal"><span class="pre">PhyloNode.get_farthest_oldest_node()</span></code></a> and
<a class="reference internal" href="../reference/reference_phylo.html#ete3.PhyloNode.get_farthest_oldest_leaf" title="ete3.PhyloNode.get_farthest_oldest_leaf"><code class="xref py py-func docutils literal"><span class="pre">PhyloNode.get_farthest_oldest_leaf()</span></code></a> can be used for that
purpose.</p>
</div>
<div class="section" id="working-with-duplicated-gene-families">
<h2><a class="toc-backref" href="#id16">Working with duplicated gene families</a><a class="headerlink" href="#working-with-duplicated-gene-families" title="Permalink to this headline">Â¶</a></h2>
<div class="section" id="treeko-splitting-gene-trees-into-species-trees">
<span id="treeko-trees"></span><h3><a class="toc-backref" href="#id17">Treeko (splitting gene trees into species trees)</a><a class="headerlink" href="#treeko-splitting-gene-trees-into-species-trees" title="Permalink to this headline">Â¶</a></h3>
<p>Comparisons between tree topologies provide important information for
many evolutionary studies. <a class="reference external" href="http://treeko.cgenomics.org">Treeko</a>
(<a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/21335609">Marcet and Gabaldon, 2011</a> ) is a novel method
that allows the comparison of any two tree topologies, even those with
missing leaves and duplications. This is important in genome-wide
analysis since many trees do not have exact leaf pairings and
therefore most tree comparison methods are rendered useless.</p>
<p>Although Treeko is available as a standalone package, it uses ETE to
generate all possible species tree topologies within a duplicated gene
family tree.</p>
<p>Thus, the ETE method <a class="reference internal" href="../reference/reference_phylo.html#ete3.PhyloNode.get_speciation_trees" title="ete3.PhyloNode.get_speciation_trees"><code class="xref py py-func docutils literal"><span class="pre">PhyloNode.get_speciation_trees()</span></code></a> is
expected to provide the core functionality required to perform a
Treeko analysis. When used, the method will return a list of all
possible species trees observed after combining the different
non-duplicated subparts under a gene family tree node.</p>
<p>Duplication events will be automatically identified using the species
overlap algorithm described within this manual. However, duplication
nodes can be manually labeled and used by disabling the
<code class="xref py py-attr docutils literal"><span class="pre">autodetect_duplication</span></code> flag.</p>
<p>Because of the combinatorial background of the Treeko method, the
number of speciation trees generated by this function may vary
enormously (ranging from few hundreds to tens of thousands topologies).</p>
<p>Here is a basic example on how to use it:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">ete3</span> <span class="kn">import</span> <span class="n">PhyloTree</span>
<span class="n">t</span> <span class="o">=</span> <span class="n">PhyloTree</span><span class="p">(</span><span class="s">&quot;((((Human_1, Chimp_1), (Human_2, (Chimp_2, Chimp_3))), ((Fish_1, (Human_3, Fish_3)), Yeast_2)), Yeast_1);&quot;</span><span class="p">)</span>
<span class="n">t</span><span class="o">.</span><span class="n">set_species_naming_function</span><span class="p">(</span><span class="k">lambda</span> <span class="n">node</span><span class="p">:</span> <span class="n">node</span><span class="o">.</span><span class="n">name</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&quot;_&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="p">)</span>

<span class="k">print</span> <span class="n">t</span><span class="o">.</span><span class="n">get_ascii</span><span class="p">(</span><span class="n">attributes</span><span class="o">=</span><span class="p">[</span><span class="s">&quot;name&quot;</span><span class="p">,</span> <span class="s">&quot;species&quot;</span><span class="p">],</span> <span class="n">show_internal</span><span class="o">=</span><span class="bp">False</span> <span class="p">)</span>

<span class="c">#            /-Human_1, Human</span>
<span class="c">#          /-|</span>
<span class="c">#         |   \-Chimp_1, Chimp</span>
<span class="c">#       /-|</span>
<span class="c">#      |  |   /-Human_2, Human</span>
<span class="c">#      |   \-|</span>
<span class="c">#      |     |   /-Chimp_2, Chimp</span>
<span class="c">#      |      \-|</span>
<span class="c">#    /-|         \-Chimp_3, Chimp</span>
<span class="c">#   |  |</span>
<span class="c">#   |  |      /-Fish_1, Fish</span>
<span class="c">#   |  |   /-|</span>
<span class="c">#   |  |  |  |   /-Human_3, Human</span>
<span class="c"># --|   \-|   \-|</span>
<span class="c">#   |     |      \-Fish_3, Fish</span>
<span class="c">#   |     |</span>
<span class="c">#   |      \-Yeast_2, Yeast</span>
<span class="c">#   |</span>
<span class="c">#    \-Yeast_1, Yeast</span>

<span class="c"># We obtain a list of species trees inferred from the duplication</span>
<span class="c"># events. Note that species specific duplications are ignored.</span>

<span class="n">ntrees</span><span class="p">,</span> <span class="n">ndups</span><span class="p">,</span> <span class="n">sptrees</span> <span class="o">=</span>  <span class="n">t</span><span class="o">.</span><span class="n">get_speciation_trees</span><span class="p">()</span>
<span class="k">print</span> <span class="s">&quot;Found </span><span class="si">%d</span><span class="s"> species trees and </span><span class="si">%d</span><span class="s"> duplication nodes&quot;</span> <span class="o">%</span><span class="p">(</span><span class="n">ntrees</span><span class="p">,</span> <span class="n">ndups</span><span class="p">)</span>
<span class="k">for</span> <span class="n">spt</span> <span class="ow">in</span> <span class="n">sptrees</span><span class="p">:</span>
   <span class="k">print</span> <span class="n">spt</span>

<span class="c"># Found 5 species trees and 4 duplication nodes</span>
<span class="c">#</span>
<span class="c">#    /-Human_1</span>
<span class="c"># --|</span>
<span class="c">#    \-Chimp_1</span>
<span class="c">#</span>
<span class="c">#    /-Human_2</span>
<span class="c"># --|</span>
<span class="c">#   |   /-Chimp_2</span>
<span class="c">#    \-|</span>
<span class="c">#       \-Chimp_3</span>
<span class="c">#</span>
<span class="c">#    /-Fish_1</span>
<span class="c"># --|</span>
<span class="c">#    \-Yeast_2</span>
<span class="c">#</span>
<span class="c">#       /-Human_3</span>
<span class="c">#    /-|</span>
<span class="c"># --|   \-Fish_3</span>
<span class="c">#   |</span>
<span class="c">#    \-Yeast_2</span>
<span class="c">#</span>
<span class="c"># --Yeast_1</span>
</pre></div>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">For performance reasons, species trees are created without any link
to the original gene family tree, rather than the species name of
each node. However, the <code class="xref py py-attr docutils literal"><span class="pre">map_features</span></code> attribute can be used
to keep certain attributes of the original tree into the generated
species trees.</p>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">Although the efficiency of the method to generate all possible
trees has been significantly improved from ETE version 2.2,
creating thousands of new PhyloTree objects could affect
performance. The flag <code class="xref py py-attr docutils literal"><span class="pre">newick_only</span></code> is now available to limit
the output to a newick string per generated tree, thus improving
the speed they can be processed or dumped into a file.</p>
</div>
</div>
<div class="section" id="splitting-gene-trees-by-duplication-events">
<span id="split-by-dup"></span><h3><a class="toc-backref" href="#id18">Splitting gene trees by duplication events</a><a class="headerlink" href="#splitting-gene-trees-by-duplication-events" title="Permalink to this headline">Â¶</a></h3>
<p>A much simpler approach to separate duplicates within the same gene
family tree is to split the topology by their duplication nodes. For
this, the method <a class="reference internal" href="../reference/reference_phylo.html#ete3.PhyloNode.split_by_dups" title="ete3.PhyloNode.split_by_dups"><code class="xref py py-func docutils literal"><span class="pre">PhyloNode.split_by_dups()</span></code></a> is provided.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">ete3</span> <span class="kn">import</span> <span class="n">PhyloTree</span>
<span class="n">t</span> <span class="o">=</span> <span class="n">PhyloTree</span><span class="p">(</span><span class="s">&quot;((((Human_1, Chimp_1), (Human_2, (Chimp_2, Chimp_3))), ((Fish_1, (Human_3, Fish_3)), Yeast_2)), Yeast_1);&quot;</span><span class="p">)</span>
<span class="n">t</span><span class="o">.</span><span class="n">set_species_naming_function</span><span class="p">(</span><span class="k">lambda</span> <span class="n">node</span><span class="p">:</span> <span class="n">node</span><span class="o">.</span><span class="n">name</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&quot;_&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="p">)</span>

<span class="k">print</span> <span class="n">t</span><span class="o">.</span><span class="n">get_ascii</span><span class="p">(</span><span class="n">attributes</span><span class="o">=</span><span class="p">[</span><span class="s">&quot;name&quot;</span><span class="p">,</span> <span class="s">&quot;species&quot;</span><span class="p">],</span> <span class="n">show_internal</span><span class="o">=</span><span class="bp">False</span> <span class="p">)</span>

<span class="c">#            /-Human_1, Human</span>
<span class="c">#          /-|</span>
<span class="c">#         |   \-Chimp_1, Chimp</span>
<span class="c">#       /-|</span>
<span class="c">#      |  |   /-Human_2, Human</span>
<span class="c">#      |   \-|</span>
<span class="c">#      |     |   /-Chimp_2, Chimp</span>
<span class="c">#      |      \-|</span>
<span class="c">#    /-|         \-Chimp_3, Chimp</span>
<span class="c">#   |  |</span>
<span class="c">#   |  |      /-Fish_1, Fish</span>
<span class="c">#   |  |   /-|</span>
<span class="c">#   |  |  |  |   /-Human_3, Human</span>
<span class="c"># --|   \-|   \-|</span>
<span class="c">#   |     |      \-Fish_3, Fish</span>
<span class="c">#   |     |</span>
<span class="c">#   |      \-Yeast_2, Yeast</span>
<span class="c">#   |</span>
<span class="c">#    \-Yeast_1, Yeast</span>

<span class="c"># Again, species specific duplications are ignored</span>
<span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="n">t</span><span class="o">.</span><span class="n">split_by_dups</span><span class="p">():</span>
    <span class="k">print</span> <span class="n">node</span>

<span class="c">#    /-Human_1</span>
<span class="c"># --|</span>
<span class="c">#    \-Chimp_1</span>
<span class="c">#</span>
<span class="c">#    /-Human_2</span>
<span class="c"># --|</span>
<span class="c">#   |   /-Chimp_2</span>
<span class="c">#    \-|</span>
<span class="c">#       \-Chimp_3</span>
<span class="c">#</span>
<span class="c"># --Yeast_2</span>
<span class="c">#</span>
<span class="c"># --Fish_1</span>
<span class="c">#</span>
<span class="c">#    /-Human_3</span>
<span class="c"># --|</span>
<span class="c">#    \-Fish_3</span>
<span class="c">#</span>
<span class="c"># --Yeast_1</span>
</pre></div>
</div>
</div>
<div class="section" id="collapse-species-specific-duplications">
<span id="collapse-expansions"></span><h3><a class="toc-backref" href="#id19">Collapse species specific duplications</a><a class="headerlink" href="#collapse-species-specific-duplications" title="Permalink to this headline">Â¶</a></h3>
<p>The method <a class="reference internal" href="../reference/reference_phylo.html#ete3.PhyloNode.collapse_lineage_specific_expansions" title="ete3.PhyloNode.collapse_lineage_specific_expansions"><code class="xref py py-func docutils literal"><span class="pre">PhyloNode.collapse_lineage_specific_expansions()</span></code></a>
method, which returns a pruned version of a tree, where nodes
representing lineage specific expansions are converted into a single
leaf node is also available.</p>
<p>From the previous examples, the lineage specific duplication of
Chimp_1 and Chimp_2 could be easily collapsed into a single node.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">ete3</span> <span class="kn">import</span> <span class="n">PhyloTree</span>
<span class="n">t</span> <span class="o">=</span> <span class="n">PhyloTree</span><span class="p">(</span><span class="s">&quot;((((Human_1, Chimp_1), (Human_2, (Chimp_2, Chimp_3))), ((Fish_1, (Human_3, Fish_3)), Yeast_2)), Yeast_1);&quot;</span><span class="p">)</span>
<span class="n">t</span><span class="o">.</span><span class="n">set_species_naming_function</span><span class="p">(</span><span class="k">lambda</span> <span class="n">node</span><span class="p">:</span> <span class="n">node</span><span class="o">.</span><span class="n">name</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&quot;_&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="p">)</span>

<span class="k">print</span> <span class="n">t</span><span class="o">.</span><span class="n">get_ascii</span><span class="p">(</span><span class="n">attributes</span><span class="o">=</span><span class="p">[</span><span class="s">&quot;name&quot;</span><span class="p">,</span> <span class="s">&quot;species&quot;</span><span class="p">],</span> <span class="n">show_internal</span><span class="o">=</span><span class="bp">False</span> <span class="p">)</span>

<span class="c">#            /-Human_1, Human</span>
<span class="c">#          /-|</span>
<span class="c">#         |   \-Chimp_1, Chimp</span>
<span class="c">#       /-|</span>
<span class="c">#      |  |   /-Human_2, Human</span>
<span class="c">#      |   \-|</span>
<span class="c">#      |     |   /-Chimp_2, Chimp</span>
<span class="c">#      |      \-|</span>
<span class="c">#    /-|         \-Chimp_3, Chimp</span>
<span class="c">#   |  |</span>
<span class="c">#   |  |      /-Fish_1, Fish</span>
<span class="c">#   |  |   /-|</span>
<span class="c">#   |  |  |  |   /-Human_3, Human</span>
<span class="c"># --|   \-|   \-|</span>
<span class="c">#   |     |      \-Fish_3, Fish</span>
<span class="c">#   |     |</span>
<span class="c">#   |      \-Yeast_2, Yeast</span>
<span class="c">#   |</span>
<span class="c">#    \-Yeast_1, Yeast</span>

<span class="n">t2</span> <span class="o">=</span> <span class="n">t</span><span class="o">.</span><span class="n">collapse_lineage_specific_expansions</span><span class="p">()</span>
<span class="k">print</span> <span class="n">t2</span><span class="o">.</span><span class="n">get_ascii</span><span class="p">(</span><span class="n">attributes</span><span class="o">=</span><span class="p">[</span><span class="s">&quot;name&quot;</span><span class="p">,</span> <span class="s">&quot;species&quot;</span><span class="p">],</span> <span class="n">show_internal</span><span class="o">=</span><span class="bp">False</span> <span class="p">)</span>

<span class="c">#             /-Human_1, Human</span>
<span class="c">#          /-|</span>
<span class="c">#         |   \-Chimp_1, Chimp</span>
<span class="c">#       /-|</span>
<span class="c">#      |  |   /-Human_2, Human</span>
<span class="c">#      |   \-|</span>
<span class="c">#      |      \-Chimp_2, Chimp   ***</span>
<span class="c">#    /-|</span>
<span class="c">#   |  |      /-Fish_1, Fish</span>
<span class="c">#   |  |   /-|</span>
<span class="c">#   |  |  |  |   /-Human_3, Human</span>
<span class="c"># --|   \-|   \-|</span>
<span class="c">#   |     |      \-Fish_3, Fish</span>
<span class="c">#   |     |</span>
<span class="c">#   |      \-Yeast_2, Yeast</span>
<span class="c">#   |</span>
<span class="c">#    \-Yeast_1, Yeast</span>
</pre></div>
</div>
</div>
</div>
</div>


           </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="tutorial_clustering.html" class="btn btn-neutral float-right" title="Clustering Trees" accesskey="n">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="tutorial_drawing.html" class="btn btn-neutral" title="The Programmable Tree Drawing Engine" accesskey="p"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2015, The ETE Toolkit Team.
      Last updated on Mar 18, 2016.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'../',
            VERSION:'3.0.0b35',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true
        };
    </script>
      <script type="text/javascript" src="../_static/jquery.js"></script>
      <script type="text/javascript" src="../_static/underscore.js"></script>
      <script type="text/javascript" src="../_static/doctools.js"></script>
      <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>

  

  
  
    <script type="text/javascript" src="../_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>