

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>PhyloTree class &mdash; ETE Toolkit - analysis and visualization of trees</title>
  

  
  
    <link rel="shortcut icon" href="../_static/favicon.ico"/>
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  

  

  
        <link rel="author" title="About these documents"
              href="../about.html"/>
    <link rel="top" title="ETE Toolkit - analysis and visualization of trees" href="../index.html"/>
        <link rel="up" title="ETE’s Reference Guide" href="index.html"/>
        <link rel="next" title="Clustering module" href="reference_clustering.html"/>
        <link rel="prev" title="Treeview module" href="reference_treeview.html"/>
<script>
    (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
           (i[r].q=i[r].q||[]).push(arguments)
    },i[r].l=1*new Date();a=s.createElement(o),
    m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
    })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

 ga('create', 'UA-783841-4', 'etetoolkit.org');
 ga('send', 'pageview');
 </script>

<!-- 
 <script type="text/javascript">
 var _gaq = _gaq || [];
 _gaq.push(['_setAccount', 'UA-783841-12']);
 _gaq.push(['_setCookiePath', '/ete2/']);
 _gaq.push(['_trackPageview']);

 (function() {
 var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
 ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
 var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
 })();
 </script>
 -->
<!-- <link rel="canonical" href="http://etetoolkit.org/docs/2.3/CANONICALPATH" />
<meta http-equiv="refresh" content="0; url=http:/etetoolkit.org/CANONICALPATH" />
<script type="text/javascript">
window.location.replace('http://etetoolkit.org/docs/2.3/' + window.location.pathname.replace('/ete2/',''));
</script> -->
 
 


  
  <script src="../_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="../index.html" class="icon icon-home"> ETE Toolkit
          

          
          </a>

          
            
            
              <div class="version">
                3.0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
                <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../changelog/index.html">Changelog history</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorial/index.html">The ETE tutorial</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">ETE&#8217;s Reference Guide</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="reference_tree.html">Master Tree class</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_treeview.html">Treeview module</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="">PhyloTree class</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_clustering.html">Clustering module</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_nexml.html">Nexml module</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_phyloxml.html">Phyloxml Module</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_seqgroup.html">Seqgroup class</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_webplugin.html">WebTreeApplication object</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_evoltree.html">EvolTree class</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_ncbi.html">NCBITaxa class</a></li>
</ul>
</li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="../index.html">ETE Toolkit</a>
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          





<div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="../index.html">Docs</a> &raquo;</li>
      
          <li><a href="index.html">ETE&#8217;s Reference Guide</a> &raquo;</li>
      
    <li>PhyloTree class</li>
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="../_sources/reference/reference_phylo.txt" rel="nofollow"> View page source</a>
          
        
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <span class="target" id="module-ete3.phylo"></span><div class="section" id="phylotree-class">
<h1>PhyloTree class<a class="headerlink" href="#phylotree-class" title="Permalink to this headline">¶</a></h1>
<dl class="class">
<dt id="ete3.PhyloNode">
<em class="property">class </em><code class="descname">PhyloNode</code><span class="sig-paren">(</span><em>newick=None</em>, <em>alignment=None</em>, <em>alg_format='fasta'</em>, <em>sp_naming_function=&lt;function _parse_species&gt;</em>, <em>format=0</em>, <em>**kargs</em><span class="sig-paren">)</span><a class="headerlink" href="#ete3.PhyloNode" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">ete3.coretype.tree.TreeNode</span></code></p>
<p>Extends the standard <a class="reference internal" href="reference_tree.html#ete3.TreeNode" title="ete3.TreeNode"><code class="xref py py-class docutils literal"><span class="pre">TreeNode</span></code></a> instance. It adds
specific attributes and methods to work with phylogentic trees.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>newick</strong> &#8211; Path to the file containing the tree or, alternatively,
the text string containing the same information.</li>
<li><strong>alignment</strong> &#8211; file containing a multiple sequence alignment.</li>
<li><strong>alg_format</strong> &#8211; &#8220;fasta&#8221;, &#8220;phylip&#8221; or &#8220;iphylip&#8221; (interleaved)</li>
<li><strong>format</strong> &#8211; <p>sub-newick format</p>
<table border="1" class="docutils">
<colgroup>
<col width="12%" />
<col width="88%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">FORMAT</th>
<th class="head">DESCRIPTION</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>0</td>
<td>flexible with support values</td>
</tr>
<tr class="row-odd"><td>1</td>
<td>flexible with internal node names</td>
</tr>
<tr class="row-even"><td>2</td>
<td>all branches + leaf names + internal supports</td>
</tr>
<tr class="row-odd"><td>3</td>
<td>all branches + all names</td>
</tr>
<tr class="row-even"><td>4</td>
<td>leaf branches + leaf names</td>
</tr>
<tr class="row-odd"><td>5</td>
<td>internal and leaf branches + leaf names</td>
</tr>
<tr class="row-even"><td>6</td>
<td>internal branches + leaf names</td>
</tr>
<tr class="row-odd"><td>7</td>
<td>leaf branches + all names</td>
</tr>
<tr class="row-even"><td>8</td>
<td>all names</td>
</tr>
<tr class="row-odd"><td>9</td>
<td>leaf names</td>
</tr>
<tr class="row-even"><td>100</td>
<td>topology only</td>
</tr>
</tbody>
</table>
</li>
<li><strong>sp_naming_function</strong> &#8211; Pointer to a parsing python
function that receives nodename as first argument and returns
the species name (see
<a class="reference internal" href="#ete3.PhyloNode.set_species_naming_function" title="ete3.PhyloNode.set_species_naming_function"><code class="xref py py-func docutils literal"><span class="pre">PhyloNode.set_species_naming_function()</span></code></a>. By default, the
3 first letter of nodes will be used as species identifiers.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">a tree node object which represents the base of the tree.</p>
</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="ete3.PhyloNode.annotate_ncbi_taxa">
<code class="descname">annotate_ncbi_taxa</code><span class="sig-paren">(</span><em>taxid_attr='species'</em>, <em>tax2name=None</em>, <em>tax2track=None</em>, <em>tax2rank=None</em>, <em>dbfile=None</em><span class="sig-paren">)</span><a class="headerlink" href="#ete3.PhyloNode.annotate_ncbi_taxa" title="Permalink to this definition">¶</a></dt>
<dd><p>Add NCBI taxonomy annotation to all descendant nodes. Leaf nodes are
expected to contain a feature (name, by default) encoding a valid taxid
number.</p>
<p>All descendant nodes (including internal nodes) are annotated with the
following new features:</p>
<p><cite>Node.spname</cite>: scientific spcies name as encoded in the NCBI taxonomy database</p>
<p><cite>Node.named_lineage</cite>: the NCBI lineage track using scientific names</p>
<p><cite>Node.taxid</cite>: NCBI taxid number</p>
<p><cite>Node.lineage</cite>: same as named_lineage but using taxid codes.</p>
<p>Note that for internal nodes, NCBI information will refer to the first
common lineage of the grouped species.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>taxid_attr</strong> (<em>name</em>) &#8211; the name of the feature that should be used to access the taxid number associated to each node.</li>
<li><strong>tax2name</strong> (<em>None</em>) &#8211; A dictionary where keys are taxid numbers and</li>
</ul>
</td>
</tr>
</tbody>
</table>
<p>values are their translation into NCBI scientific name. Its use is
optional and allows to avoid database queries when annotating many trees
containing the same set of taxids.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>tax2track</strong> (<em>None</em>) &#8211; A dictionary where keys are taxid numbers and</td>
</tr>
</tbody>
</table>
<p>values are their translation into NCBI lineage tracks (taxids). Its use is
optional and allows to avoid database queries when annotating many trees
containing the same set of taxids.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>tax2rank</strong> (<em>None</em>) &#8211; A dictionary where keys are taxid numbers and</td>
</tr>
</tbody>
</table>
<p>values are their translation into NCBI rank name. Its use is optional
and allows to avoid database queries when annotating many trees
containing the same set of taxids.</p>
<p>:param None dbfile : If provided, the provided file will be used as a
local copy of the NCBI taxonomy database.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">tax2name (a dictionary translating taxid numbers into</td>
</tr>
</tbody>
</table>
<p>scientific name), tax2lineage (a dictionary translating taxid numbers
into their corresponding NCBI lineage track) and tax2rank (a dictionary translating taxid numbers into
rank names).</p>
</dd></dl>

<dl class="method">
<dt id="ete3.PhyloNode.collapse_lineage_specific_expansions">
<code class="descname">collapse_lineage_specific_expansions</code><span class="sig-paren">(</span><em>species=None</em>, <em>return_copy=True</em><span class="sig-paren">)</span><a class="headerlink" href="#ete3.PhyloNode.collapse_lineage_specific_expansions" title="Permalink to this definition">¶</a></dt>
<dd><p>Converts lineage specific expansion nodes into a single
tip node (randomly chosen from tips within the expansion).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>species</strong> (<em>None</em>) &#8211; If supplied, only expansions matching the
species criteria will be pruned. When None, all expansions
within the tree will be processed.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="ete3.PhyloNode.get_age">
<code class="descname">get_age</code><span class="sig-paren">(</span><em>species2age</em><span class="sig-paren">)</span><a class="headerlink" href="#ete3.PhyloNode.get_age" title="Permalink to this definition">¶</a></dt>
<dd><p>Implements the phylostratigrafic method described in:</p>
<p>Huerta-Cepas, J., &amp; Gabaldon, T. (2011). Assigning duplication events to
relative temporal scales in genome-wide studies. Bioinformatics, 27(1),
38-45.</p>
</dd></dl>

<dl class="method">
<dt id="ete3.PhyloNode.get_age_balanced_outgroup">
<code class="descname">get_age_balanced_outgroup</code><span class="sig-paren">(</span><em>species2age</em><span class="sig-paren">)</span><a class="headerlink" href="#ete3.PhyloNode.get_age_balanced_outgroup" title="Permalink to this definition">¶</a></dt>
<dd><div class="versionadded">
<p><span class="versionmodified">New in version 2.2.</span></p>
</div>
<p>Returns the node better balance current tree structure
according to the topological age of the different leaves and
internal node sizes.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>species2age</strong> &#8211; A dictionary translating from leaf names
into a topological age.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="ete3.PhyloNode.get_descendant_evol_events">
<code class="descname">get_descendant_evol_events</code><span class="sig-paren">(</span><em>sos_thr=0.0</em><span class="sig-paren">)</span><a class="headerlink" href="#ete3.PhyloNode.get_descendant_evol_events" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a list of <strong>all</strong> duplication and speciation
events detected after this node. Nodes are assumed to be
duplications when a species overlap is found between its child
linages. Method is described more detail in:</p>
<p>&#8220;The Human Phylome.&#8221; Huerta-Cepas J, Dopazo H, Dopazo J, Gabaldon
T. Genome Biol. 2007;8(6):R109.</p>
</dd></dl>

<dl class="method">
<dt id="ete3.PhyloNode.get_farthest_oldest_leaf">
<code class="descname">get_farthest_oldest_leaf</code><span class="sig-paren">(</span><em>species2age</em>, <em>is_leaf_fn=None</em><span class="sig-paren">)</span><a class="headerlink" href="#ete3.PhyloNode.get_farthest_oldest_leaf" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the farthest oldest leaf to the current
one. It requires an species2age dictionary with the age
estimation for all species.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>is_leaf_fn</strong> (<em>None</em>) &#8211; A pointer to a function that
receives a node instance as unique argument and returns True
or False. It can be used to dynamically collapse nodes, so
they are seen as leaves.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="ete3.PhyloNode.get_farthest_oldest_node">
<code class="descname">get_farthest_oldest_node</code><span class="sig-paren">(</span><em>species2age</em><span class="sig-paren">)</span><a class="headerlink" href="#ete3.PhyloNode.get_farthest_oldest_node" title="Permalink to this definition">¶</a></dt>
<dd><div class="versionadded">
<p><span class="versionmodified">New in version 2.1.</span></p>
</div>
<p>Returns the farthest oldest node (leaf or internal). The
difference with get_farthest_oldest_leaf() is that in this
function internal nodes grouping seqs from the same species
are collapsed.</p>
</dd></dl>

<dl class="method">
<dt id="ete3.PhyloNode.get_my_evol_events">
<code class="descname">get_my_evol_events</code><span class="sig-paren">(</span><em>sos_thr=0.0</em><span class="sig-paren">)</span><a class="headerlink" href="#ete3.PhyloNode.get_my_evol_events" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a list of duplication and speciation events in
which the current node has been involved. Scanned nodes are
also labeled internally as dup=True|False. You can access this
labels using the &#8216;node.dup&#8217; sintaxis.</p>
<p>Method: the algorithm scans all nodes from the given leafName to
the root. Nodes are assumed to be duplications when a species
overlap is found between its child linages. Method is described
more detail in:</p>
<p>&#8220;The Human Phylome.&#8221; Huerta-Cepas J, Dopazo H, Dopazo J, Gabaldon
T. Genome Biol. 2007;8(6):R109.</p>
</dd></dl>

<dl class="method">
<dt id="ete3.PhyloNode.get_speciation_trees">
<code class="descname">get_speciation_trees</code><span class="sig-paren">(</span><em>map_features=None</em>, <em>autodetect_duplications=True</em>, <em>newick_only=False</em>, <em>target_attr='species'</em><span class="sig-paren">)</span><a class="headerlink" href="#ete3.PhyloNode.get_speciation_trees" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates all possible species trees contained within a
duplicated gene family tree as described in <a class="reference external" href="http://treeko.cgenomics.org">Treeko</a> (see <a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/21335609">Marcet and Gabaldon,
2011</a> ).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>autodetect_duplications</strong> (<em>True</em>) &#8211; If True, duplication</td>
</tr>
</tbody>
</table>
<p>nodes will be automatically detected using the Species Overlap
algorithm (<code class="xref py py-func docutils literal"><span class="pre">PhyloNode.get_descendants_evol_events()</span></code>. If
False, duplication nodes within the original tree are expected
to contain the feature &#8220;evoltype=D&#8221;.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>features</strong> (<em>None</em>) &#8211; A list of features that should be</td>
</tr>
</tbody>
</table>
<p>mapped from the original gene family tree to each species
tree subtree.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">(number_of_sptrees, number_of_dups, species_tree_iterator)</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="ete3.PhyloNode.get_species">
<code class="descname">get_species</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#ete3.PhyloNode.get_species" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the set of species covered by its partition.</p>
</dd></dl>

<dl class="method">
<dt id="ete3.PhyloNode.iter_species">
<code class="descname">iter_species</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#ete3.PhyloNode.iter_species" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns an iterator over the species grouped by this node.</p>
</dd></dl>

<dl class="method">
<dt id="ete3.PhyloNode.link_to_alignment">
<code class="descname">link_to_alignment</code><span class="sig-paren">(</span><em>alignment</em>, <em>alg_format='fasta'</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#ete3.PhyloNode.link_to_alignment" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="ete3.PhyloNode.ncbi_compare">
<code class="descname">ncbi_compare</code><span class="sig-paren">(</span><em>autodetect_duplications=True</em>, <em>cached_content=None</em><span class="sig-paren">)</span><a class="headerlink" href="#ete3.PhyloNode.ncbi_compare" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="ete3.PhyloNode.reconcile">
<code class="descname">reconcile</code><span class="sig-paren">(</span><em>species_tree</em><span class="sig-paren">)</span><a class="headerlink" href="#ete3.PhyloNode.reconcile" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the reconcilied topology with the provided species
tree, and a list of evolutionary events inferred from such
reconciliation.</p>
</dd></dl>

<dl class="method">
<dt id="ete3.PhyloNode.set_species_naming_function">
<code class="descname">set_species_naming_function</code><span class="sig-paren">(</span><em>fn</em><span class="sig-paren">)</span><a class="headerlink" href="#ete3.PhyloNode.set_species_naming_function" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets the parsing function used to extract species name from a
node&#8217;s name.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>fn</strong> &#8211; Pointer to a parsing python function that
receives nodename as first argument and returns the species
name.</td>
</tr>
</tbody>
</table>
<div class="highlight-python"><div class="highlight"><pre><span class="c"># Example of a parsing function to extract species names for</span>
<span class="c"># all nodes in a given tree.</span>
<span class="k">def</span> <span class="nf">parse_sp_name</span><span class="p">(</span><span class="n">node_name</span><span class="p">):</span>
    <span class="k">return</span> <span class="n">node_name</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&quot;_&quot;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span>
<span class="n">tree</span><span class="o">.</span><span class="n">set_species_naming_function</span><span class="p">(</span><span class="n">parse_sp_name</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="attribute">
<dt id="ete3.PhyloNode.species">
<code class="descname">species</code><a class="headerlink" href="#ete3.PhyloNode.species" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="ete3.PhyloNode.split_by_dups">
<code class="descname">split_by_dups</code><span class="sig-paren">(</span><em>autodetect_duplications=True</em><span class="sig-paren">)</span><a class="headerlink" href="#ete3.PhyloNode.split_by_dups" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the list of all subtrees resulting from splitting
current tree by its duplication nodes.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>autodetect_duplications</strong> (<em>True</em>) &#8211; If True, duplication</td>
</tr>
</tbody>
</table>
<p>nodes will be automatically detected using the Species Overlap
algorithm (<code class="xref py py-func docutils literal"><span class="pre">PhyloNode.get_descendants_evol_events()</span></code>. If
False, duplication nodes within the original tree are expected
to contain the feature &#8220;evoltype=D&#8221;.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">species_trees</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="attribute">
<dt id="ete3.PhyloTree">
<code class="descname">PhyloTree</code><a class="headerlink" href="#ete3.PhyloTree" title="Permalink to this definition">¶</a></dt>
<dd><p>alias of <a class="reference internal" href="#ete3.PhyloNode" title="ete3.PhyloNode"><code class="xref py py-class docutils literal"><span class="pre">PhyloNode</span></code></a></p>
</dd></dl>

<dl class="class">
<dt id="ete3.phylo.EvolEvent">
<em class="property">class </em><code class="descname">EvolEvent</code><a class="headerlink" href="#ete3.phylo.EvolEvent" title="Permalink to this definition">¶</a></dt>
<dd><p>Basic evolutionary event. It stores all the information about an
event(node) ocurred in a phylogenetic tree.</p>
<p><code class="xref py py-attr docutils literal"><span class="pre">etype</span></code> : <code class="docutils literal"><span class="pre">D</span></code> (Duplication), <code class="docutils literal"><span class="pre">S</span></code> (Speciation), <code class="docutils literal"><span class="pre">L</span></code> (gene loss),</p>
<p><code class="xref py py-attr docutils literal"><span class="pre">in_seqs</span></code> : the list of sequences in one side of the event.</p>
<p><code class="xref py py-attr docutils literal"><span class="pre">out_seqs</span></code> : the list of sequences in the other side of the event</p>
<p><code class="xref py py-attr docutils literal"><span class="pre">node</span></code> : link to the event node in the tree</p>
</dd></dl>

</div>


           </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="reference_clustering.html" class="btn btn-neutral float-right" title="Clustering module" accesskey="n">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="reference_treeview.html" class="btn btn-neutral" title="Treeview module" accesskey="p"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2015, The ETE Toolkit Team.
      Last updated on Mar 18, 2016.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'../',
            VERSION:'3.0.0b35',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true
        };
    </script>
      <script type="text/javascript" src="../_static/jquery.js"></script>
      <script type="text/javascript" src="../_static/underscore.js"></script>
      <script type="text/javascript" src="../_static/doctools.js"></script>
      <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>

  

  
  
    <script type="text/javascript" src="../_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>